{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "little-radical",
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp datasets.OVPDataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "organic-argument",
   "metadata": {},
   "source": [
    "# OVP dataset\n",
    "\n",
    "> API details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "higher-salmon",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *\n",
    "from corradin_ovp_utils.catalog import test_data_catalog, conf_test_data_catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "resident-playlist",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from typing import Any, Dict, List, Optional, Literal, Union\n",
    "from enum import Enum\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from kedro.io import AbstractVersionedDataSet\n",
    "from pydantic import BaseModel\n",
    "from pydantic.dataclasses import dataclass\n",
    "from dataclasses import InitVar, asdict\n",
    "\n",
    "from kedro.io.core import (\n",
    "    AbstractVersionedDataSet,\n",
    "    DataSetError,\n",
    "    Version,\n",
    "    get_filepath_str,\n",
    "    get_protocol_and_path,\n",
    ")\n",
    "import fsspec\n",
    "from copy import deepcopy\n",
    "from pathlib import Path, PosixPath\n",
    "from types import SimpleNamespace\n",
    "from fastcore.meta import delegates\n",
    "import inspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fatal-cowboy",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "from corradin_ovp_utils.datasets import genetic_file, sample_file\n",
    "from corradin_ovp_utils.datasets.schemas import SingleFilePathSchema, CaseControlFilePathSchema\n",
    "import corradin_ovp_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "awful-equality",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "#modified from kedro.io.core\n",
    "from kedro.utils import load_obj\n",
    "\n",
    "_DEFAULT_PACKAGES = [\"kedro.io.\", \"kedro.extras.datasets.\", \"corradin_ovp_utils.datasets.\", \"corradin_ovp_utils.datasets.genetic_file.Gen\", \"\"]\n",
    "\n",
    "class OVPDataSetError(Exception):\n",
    "    pass\n",
    "\n",
    "def parse_class(key,\n",
    "    class_obj:str\n",
    "): #-> Tuple[Type[AbstractDataSet], Dict[str, Any]]:\n",
    "    \"\"\"Parse and instantiate a dataset class using the configuration provided.\n",
    "    Args:\n",
    "        config: Data set config dictionary. It *must* contain the `type` key\n",
    "            with fully qualified class name.\n",
    "        load_version: Version string to be used for ``load`` operation if\n",
    "                the data set is versioned. Has no effect on the data set\n",
    "                if versioning was not enabled.\n",
    "        save_version: Version string to be used for ``save`` operation if\n",
    "            the data set is versioned. Has no effect on the data set\n",
    "            if versioning was not enabled.\n",
    "    Raises:\n",
    "        DataSetError: If the function fails to parse the configuration provided.\n",
    "    Returns:\n",
    "        2-tuple: (Dataset class object, configuration dictionary)\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(class_obj, str):\n",
    "        if len(class_obj.strip(\".\")) != len(class_obj):\n",
    "            raise OVPDataSetError(\n",
    "                f\"{key} class path does not support relative \"\n",
    "                \"paths or paths ending with a dot.\"\n",
    "            )\n",
    "\n",
    "        class_paths = (prefix + class_obj for prefix in _DEFAULT_PACKAGES)\n",
    "\n",
    "        trials = (_load_obj(class_path) for class_path in class_paths)\n",
    "        try:\n",
    "            class_obj = next(obj for obj in trials if obj is not None)\n",
    "        except StopIteration as exc:\n",
    "            raise OVPDataSetError(f\"Class `{class_obj}` not found.\") from exc\n",
    "\n",
    "#     if not issubclass(class_obj, AbstractDataSet):\n",
    "#         raise DataSetError(\n",
    "#             f\"DataSet type `{class_obj.__module__}.{class_obj.__qualname__}` \"\n",
    "#             f\"is invalid: all data set types must extend `AbstractDataSet`.\"\n",
    "#         )\n",
    "\n",
    "    return class_obj\n",
    "\n",
    "\n",
    "def _load_obj(class_path: str) -> Optional[object]:\n",
    "    mod_path, _, class_name = class_path.rpartition(\".\")\n",
    "    try:\n",
    "        available_classes = load_obj(f\"{mod_path}.__all__\")\n",
    "    # ModuleNotFoundError: When `load_obj` can't find `mod_path` (e.g `kedro.io.pandas`)\n",
    "    #                      this is because we try a combination of all prefixes.\n",
    "    # AttributeError: When `load_obj` manages to load `mod_path` but it doesn't have an\n",
    "    #                 `__all__` attribute -- either because it's a custom or a kedro.io dataset\n",
    "    except (ModuleNotFoundError, AttributeError, ValueError):\n",
    "        available_classes = None\n",
    "\n",
    "    try:\n",
    "        class_obj = load_obj(class_path)\n",
    "    except (ModuleNotFoundError, ValueError):\n",
    "        return None\n",
    "    except AttributeError as exc:\n",
    "        if available_classes and class_name in available_classes:\n",
    "            raise DataSetError(\n",
    "                f\"{exc} Please see the documentation on how to \"\n",
    "                f\"install relevant dependencies for {class_path}:\\n\"\n",
    "                f\"https://kedro.readthedocs.io/en/stable/\"\n",
    "                f\"04_kedro_project_setup/01_dependencies.html\"\n",
    "            ) from exc\n",
    "        return None\n",
    "\n",
    "    return class_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "provincial-stationery",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert parse_class(\"file_format\", \"genetic_file.Gen.GenFileFormat\") == genetic_file.GenFileFormat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cellular-photograph",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert parse_class(\"file_format\", \"corradin_ovp_utils.datasets.genetic_file.Gen.GenFileFormat\") == genetic_file.GenFileFormat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "creative-alignment",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "namespace(case=GenFileFormat(file_path=SingleFilePathSchema(folder='data/test_data/gen_file', full_file_name='test_CASE_MS_chr22.gen', file_name='test_CASE_MS_chr22', extension='gen', split_by_chromosome=None), prob_n_cols=3, initial_cols=['dashes', 'rsid', 'position', 'ref', 'alt'], rsid_col='rsid', ref_col='ref', alt_col='alt', ref_alt_delim=None, pandas_args={'sep': ' ', 'header': None}, sample_ids=None, current_file_path=None), control=GenFileFormat(file_path=SingleFilePathSchema(folder='data/test_data/gen_file', full_file_name='test_CONTROL_MS_chr22.gen', file_name='test_CONTROL_MS_chr22', extension='gen', split_by_chromosome=None), prob_n_cols=3, initial_cols=['dashes', 'rsid', 'position', 'ref', 'alt'], rsid_col='rsid', ref_col='ref', alt_col='alt', ref_alt_delim=None, pandas_args={'sep': ' ', 'header': None}, sample_ids=None, current_file_path=None))"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data_catalog.load(\"genetic_file\").files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "extraordinary-humor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "namespace(single_file=GenFileFormat(file_path=SingleFilePathSchema(folder='data/test_data/gen_file', full_file_name='test_CASE_MS_chr22.gen', file_name='test_CASE_MS_chr22', extension='gen', split_by_chromosome=None), prob_n_cols=3, initial_cols=['dashes', 'rsid', 'position', 'ref', 'alt'], rsid_col='rsid', ref_col='ref', alt_col='alt', ref_alt_delim=None, pandas_args={'sep': ' ', 'header': None}, sample_ids=None, current_file_path=None))"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data_catalog.load(\"genetic_file_single\").files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continuous-conflict",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export    \n",
    "# class FILE_FORMAT_ENUM(Enum):\n",
    "#     GenFile = genetic_datasets.GenFileFormat\n",
    "#     SampleFile = sample_file.SampleFileFormat\n",
    "    \n",
    "# class FILE_TYPE_ENUM(Enum):\n",
    "#     CC = CaseControlFilePathSchema\n",
    "#     S = SingleFilePathSchema\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "addressed-magnitude",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "class OVPDataset(AbstractVersionedDataSet):\n",
    "    def __init__(self,\n",
    "                 file_type,\n",
    "                 file_format,\n",
    "                 file_path,\n",
    "                 common_folder=None,\n",
    "                load_args: Dict[str, Any] = None,\n",
    "                version: Version = None,\n",
    "                credentials: Dict[str, Any] = None,\n",
    "                fs_args: Dict[str, Any] = None,\n",
    "                ):\n",
    "        \n",
    "        self.file_type = file_type\n",
    "        self._file_path_class = parse_class(\"file_type\", file_type)\n",
    "        self._file_path = self._file_path_class(**file_path) if common_folder is None else self._file_path_class(**file_path, common_folder = common_folder) #custom file path\n",
    "        \n",
    "        \n",
    "        self._version = version\n",
    "        \n",
    "        self._file_format_class = parse_class(\"file_format\", file_format)\n",
    "        #self._file_format = self._file_format_class(**load_args)\n",
    "        \n",
    "        if self._file_path_class != SingleFilePathSchema:\n",
    "            self.files = SimpleNamespace(**{single_file : self._file_format_class(file_path = single_file_path, **load_args)\\\n",
    "                                        for single_file, single_file_path in self.full_file_path.items()})#self._file_path.apply_func(self._file_format_class, **load_args)\n",
    "        else:\n",
    "            self.files = SimpleNamespace(**{\"single_file\" : self._file_format_class(file_path = self.full_file_path, **load_args)})\n",
    "            \n",
    "        _fs_args = deepcopy(fs_args) or {}\n",
    "        _fs_open_args_load = _fs_args.pop(\"open_args_load\", {})\n",
    "        _fs_open_args_save = _fs_args.pop(\"open_args_save\", {})\n",
    "        _credentials = deepcopy(credentials) or {}\n",
    "        \n",
    "        #protocol, path = get_protocol_and_path(filepath, version)\n",
    "        if self._file_path.protocol == \"file\":\n",
    "            _fs_args.setdefault(\"auto_mkdir\", True)\n",
    "        \n",
    "        _fs_open_args_save.setdefault(\"mode\", \"w\")\n",
    "        self._fs_open_args_load = _fs_open_args_load\n",
    "        self._fs_open_args_save = _fs_open_args_save\n",
    "        \n",
    "        self._protocol = self._file_path.protocol\n",
    "        self._fs = fsspec.filesystem(self._protocol, **_credentials, **_fs_args)\n",
    "\n",
    "#         self._protocol = protocol\n",
    "#         self._fs = fsspec.filesystem(self._protocol, **_credentials, **_fs_args)\n",
    "    \n",
    "    \n",
    "    @property\n",
    "    def full_file_path(self):\n",
    "        return self._file_path.file_path_obj\n",
    "    \n",
    "    def load_all_files(self):\n",
    "        return pd.concat([file.load() for _, file in vars(self.files).items()])\n",
    "            \n",
    "    def _load(self):\n",
    "        return self\n",
    "    \n",
    "    def _save(self):\n",
    "        pass\n",
    "    \n",
    "    def _describe(self):\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "elder-security",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'corradin_ovp_utils.datasets.OVPDataset.OVPDataset'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_conf_test_data_catalog = conf_test_data_catalog[\"genetic_file\"].copy()\n",
    "test_conf_test_data_catalog.pop(\"type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "reasonable-theorem",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.OVPDataset at 0x1497c8acf760>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_genetic_dataset = OVPDataset(**test_conf_test_data_catalog)\n",
    "test_genetic_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "detailed-factory",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_case_file = test_genetic_dataset.files.case\n",
    "test_control_file = test_genetic_dataset.files.control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "alternative-wednesday",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Invalid file path or buffer object type: <class 'corradin_ovp_utils.datasets.schemas.SingleFilePathSchema'>",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-05762c8173cf>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtest_case_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_genotypes_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py\u001b[0m in \u001b[0;36mget_genotypes_df\u001b[0;34m(self, chrom, rsid_list, **kwargs)\u001b[0m\n\u001b[1;32m    126\u001b[0m             \u001b[0mgeno_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_rsid_df\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrsid_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrsid_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0musecols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen_info_cols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    127\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 128\u001b[0;31m             \u001b[0mgeno_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mchrom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0musecols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen_info_cols\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    129\u001b[0m         \u001b[0mgeno_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"homo_ref\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgeno_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref_col\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    130\u001b[0m         make_het_geno_func = lambda row: ''.join(sorted([row[self.ref_col],\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py\u001b[0m in \u001b[0;36mload\u001b[0;34m(self, chrom, **kwargs)\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchrom\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcurrent_file_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_resolved_file_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# use another attribute so `load` and `_load_unprocessed` can share\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcurrent_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    107\u001b[0m         \u001b[0;31m#df.columns = self.column_headers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py\u001b[0m in \u001b[0;36mload_args\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     92\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload_args\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m         \u001b[0mload_args\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpandas_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 94\u001b[0;31m         \u001b[0mcol_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumn_headers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     95\u001b[0m         \u001b[0mload_args\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcol_names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mload_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py\u001b[0m in \u001b[0;36mcolumn_headers\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcolumn_headers\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m         \u001b[0;31m#read the first non-header line\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 71\u001b[0;31m         \u001b[0mfirst_line_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_load_unprocessed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchunksize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     72\u001b[0m         \u001b[0mgenerator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcol_name_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     73\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenerator\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcol\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfirst_line_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py\u001b[0m in \u001b[0;36m_load_unprocessed\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m     98\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mdelegates\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_load_unprocessed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 100\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcurrent_file_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpandas_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    101\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    603\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    812\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    813\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 814\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    815\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    816\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1043\u001b[0m             )\n\u001b[1;32m   1044\u001b[0m         \u001b[0;31m# error: Too many arguments for \"ParserBase\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1045\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore[call-arg]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1046\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1047\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_failover_to_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1860\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1861\u001b[0m         \u001b[0;31m# open handles\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1862\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open_handles\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1863\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1864\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m\"storage_options\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"encoding\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"memory_map\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"compression\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_open_handles\u001b[0;34m(self, src, kwds)\u001b[0m\n\u001b[1;32m   1355\u001b[0m         \u001b[0mLet\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mreaders\u001b[0m \u001b[0mopen\u001b[0m \u001b[0mIOHanldes\u001b[0m \u001b[0mafter\u001b[0m \u001b[0mthey\u001b[0m \u001b[0mare\u001b[0m \u001b[0mdone\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mtheir\u001b[0m \u001b[0mpotential\u001b[0m \u001b[0mraises\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1356\u001b[0m         \"\"\"\n\u001b[0;32m-> 1357\u001b[0;31m         self.handles = get_handle(\n\u001b[0m\u001b[1;32m   1358\u001b[0m             \u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1359\u001b[0m             \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    556\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    557\u001b[0m     \u001b[0;31m# open URLs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 558\u001b[0;31m     ioargs = _get_filepath_or_buffer(\n\u001b[0m\u001b[1;32m    559\u001b[0m         \u001b[0mpath_or_buf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    560\u001b[0m         \u001b[0mencoding\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/common.py\u001b[0m in \u001b[0;36m_get_filepath_or_buffer\u001b[0;34m(filepath_or_buffer, encoding, compression, mode, storage_options)\u001b[0m\n\u001b[1;32m    369\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_file_like\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    370\u001b[0m         \u001b[0mmsg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"Invalid file path or buffer object type: {type(filepath_or_buffer)}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 371\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    372\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    373\u001b[0m     return IOArgs(\n",
      "\u001b[0;31mValueError\u001b[0m: Invalid file path or buffer object type: <class 'corradin_ovp_utils.datasets.schemas.SingleFilePathSchema'>"
     ]
    }
   ],
   "source": [
    "test_case_file.get_genotypes_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "double-warner",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "  and should_run_async(code)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>homo_ref</th>\n",
       "      <th>het</th>\n",
       "      <th>homo_alt</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rsid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs77948203</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>GG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1014626</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs9610458</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs5762201</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1004237</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs134490</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4821519</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GG</td>\n",
       "      <td>CG</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1003500</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs5756405</th>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ref alt homo_ref het homo_alt\n",
       "rsid                                    \n",
       "rs77948203   G   A       GG  AG       AA\n",
       "rs1014626    C   T       CC  CT       TT\n",
       "rs9610458    C   T       CC  CT       TT\n",
       "rs5762201    A   G       AA  AG       GG\n",
       "rs1004237    C   T       CC  CT       TT\n",
       "rs134490     C   T       CC  CT       TT\n",
       "rs4821519    G   C       GG  CG       CC\n",
       "rs1003500    C   T       CC  CT       TT\n",
       "rs5756405    A   G       AA  AG       GG"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_control_file.get_genotypes_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "modified-prague",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "  and should_run_async(code)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cce7efbad4f847dd91b1645a4e98c5d1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>rsid</th>\n",
       "      <th>rs77948203</th>\n",
       "      <th>rs1014626</th>\n",
       "      <th>rs9610458</th>\n",
       "      <th>rs5762201</th>\n",
       "      <th>rs1004237</th>\n",
       "      <th>rs134490</th>\n",
       "      <th>rs4821519</th>\n",
       "      <th>rs1003500</th>\n",
       "      <th>rs5756405</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>NA</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample10</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1000</th>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1001</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample995</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample996</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample997</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample998</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample999</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9772 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "rsid       rs77948203 rs1014626 rs9610458 rs5762201 rs1004237 rs134490  \\\n",
       "sample_id                                                                \n",
       "sample1            GG        TT        TT        GG        CC       NA   \n",
       "sample10           GG        TT        TT        GG        CC       TT   \n",
       "sample100          GG        TT        TT        GG        CC       TT   \n",
       "sample1000         AG        TT        CT        GG        CC       CT   \n",
       "sample1001         GG        TT        TT        GG        CC       CT   \n",
       "...               ...       ...       ...       ...       ...      ...   \n",
       "sample995          GG        TT        CT        GG        CC       CT   \n",
       "sample996          GG        TT        CT        GG        CC       TT   \n",
       "sample997          GG        TT        CT        AG        CC       TT   \n",
       "sample998          GG        TT        CT        GG        CC       TT   \n",
       "sample999          GG        TT        CT        AG        CC       TT   \n",
       "\n",
       "rsid       rs4821519 rs1003500 rs5756405  \n",
       "sample_id                                 \n",
       "sample1           GG        CC        AG  \n",
       "sample10          GG        CC        AG  \n",
       "sample100         GG        CC        AG  \n",
       "sample1000        GG        CC        AA  \n",
       "sample1001        CG        CC        AG  \n",
       "...              ...       ...       ...  \n",
       "sample995         CC        CC        AG  \n",
       "sample996         GG        CC        AG  \n",
       "sample997         GG        CC        AG  \n",
       "sample998         GG        CC        AA  \n",
       "sample999         GG        CC        AG  \n",
       "\n",
       "[9772 rows x 9 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_all_geno_file = test_case_file.apply_func_to_all_samples(genetic_file.triplicate_converter, \n",
    "                                              genotype_df = test_case_file.get_genotypes_df()).T\n",
    "case_all_geno_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "living-grenada",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "  and should_run_async(code)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "02507d4dfa354a02966968e2051e85a9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>rsid</th>\n",
       "      <th>rs77948203</th>\n",
       "      <th>rs1014626</th>\n",
       "      <th>rs9610458</th>\n",
       "      <th>rs5762201</th>\n",
       "      <th>rs1004237</th>\n",
       "      <th>rs134490</th>\n",
       "      <th>rs4821519</th>\n",
       "      <th>rs1003500</th>\n",
       "      <th>rs5756405</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample10</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>NA</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1000</th>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1001</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample995</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample996</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample997</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>NA</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample998</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample999</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CG</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5175 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "rsid       rs77948203 rs1014626 rs9610458 rs5762201 rs1004237 rs134490  \\\n",
       "sample_id                                                                \n",
       "sample1            GG        TT        TT        AG        CC       CT   \n",
       "sample10           GG        TT        CT        GG        CC       TT   \n",
       "sample100          GG        TT        CT        GG        CC       NA   \n",
       "sample1000         AG        TT        TT        GG        CC       CT   \n",
       "sample1001         GG        TT        CC        GG        CC       CT   \n",
       "...               ...       ...       ...       ...       ...      ...   \n",
       "sample995          GG        TT        TT        GG        CC       TT   \n",
       "sample996          GG        TT        CT        GG        CC       TT   \n",
       "sample997          GG        TT        CC        GG        CC       NA   \n",
       "sample998          GG        TT        CT        GG        CC       TT   \n",
       "sample999          GG        TT        CT        GG        CC       CT   \n",
       "\n",
       "rsid       rs4821519 rs1003500 rs5756405  \n",
       "sample_id                                 \n",
       "sample1           GG        CC        AG  \n",
       "sample10          GG        CC        AA  \n",
       "sample100         GG        CC        AG  \n",
       "sample1000        GG        CC        AA  \n",
       "sample1001        GG        CC        AG  \n",
       "...              ...       ...       ...  \n",
       "sample995         GG        CC        AG  \n",
       "sample996         GG        CC        AA  \n",
       "sample997         GG        CC        AG  \n",
       "sample998         GG        CC        AG  \n",
       "sample999         CG        CC        GG  \n",
       "\n",
       "[5175 rows x 9 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_all_geno_file = test_control_file.apply_func_to_all_samples(genetic_file.triplicate_converter, \n",
    "                                              genotype_df = test_control_file.get_genotypes_df()).T\n",
    "control_all_geno_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "alive-research",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "  and should_run_async(code)\n"
     ]
    }
   ],
   "source": [
    "assert not case_all_geno_file.equals(control_all_geno_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "resident-rebound",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs77948203</th>\n",
       "      <th>rs1014626</th>\n",
       "      <th>rs9610458</th>\n",
       "      <th>rs5762201</th>\n",
       "      <th>rs1004237</th>\n",
       "      <th>rs134490</th>\n",
       "      <th>rs4821519</th>\n",
       "      <th>rs1003500</th>\n",
       "      <th>rs5756405</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample10</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1000</th>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1001</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample995</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample996</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample997</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample998</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample999</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9772 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs77948203 rs1014626 rs9610458 rs5762201 rs1004237 rs134490  \\\n",
       "sample_id                                                                \n",
       "sample1            GG        TT        TT        GG        CC      NaN   \n",
       "sample10           GG        TT        TT        GG        CC       TT   \n",
       "sample100          GG        TT        TT        GG        CC       TT   \n",
       "sample1000         AG        TT        CT        GG        CC       CT   \n",
       "sample1001         GG        TT        TT        GG        CC       CT   \n",
       "...               ...       ...       ...       ...       ...      ...   \n",
       "sample995          GG        TT        CT        GG        CC       CT   \n",
       "sample996          GG        TT        CT        GG        CC       TT   \n",
       "sample997          GG        TT        CT        AG        CC       TT   \n",
       "sample998          GG        TT        CT        GG        CC       TT   \n",
       "sample999          GG        TT        CT        AG        CC       TT   \n",
       "\n",
       "           rs4821519 rs1003500 rs5756405  \n",
       "sample_id                                 \n",
       "sample1           GG        CC        AG  \n",
       "sample10          GG        CC        AG  \n",
       "sample100         GG        CC        AG  \n",
       "sample1000        GG        CC        AA  \n",
       "sample1001        CG        CC        AG  \n",
       "...              ...       ...       ...  \n",
       "sample995         CC        CC        AG  \n",
       "sample996         GG        CC        AG  \n",
       "sample997         GG        CC        AG  \n",
       "sample998         GG        CC        AA  \n",
       "sample999         GG        CC        AG  \n",
       "\n",
       "[9772 rows x 9 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data_catalog.save(\"case_geno_each_sample\", case_all_geno_file)\n",
    "test_data_catalog.load(\"case_geno_each_sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "warming-royal",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.\n",
      "  and should_run_async(code)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs77948203</th>\n",
       "      <th>rs1014626</th>\n",
       "      <th>rs9610458</th>\n",
       "      <th>rs5762201</th>\n",
       "      <th>rs1004237</th>\n",
       "      <th>rs134490</th>\n",
       "      <th>rs4821519</th>\n",
       "      <th>rs1003500</th>\n",
       "      <th>rs5756405</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>AG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample10</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1000</th>\n",
       "      <td>AG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample1001</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample995</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample996</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample997</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample998</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>TT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample999</th>\n",
       "      <td>GG</td>\n",
       "      <td>TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>GG</td>\n",
       "      <td>CC</td>\n",
       "      <td>CT</td>\n",
       "      <td>CG</td>\n",
       "      <td>CC</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5175 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rs77948203 rs1014626 rs9610458 rs5762201 rs1004237 rs134490  \\\n",
       "sample_id                                                                \n",
       "sample1            GG        TT        TT        AG        CC       CT   \n",
       "sample10           GG        TT        CT        GG        CC       TT   \n",
       "sample100          GG        TT        CT        GG        CC      NaN   \n",
       "sample1000         AG        TT        TT        GG        CC       CT   \n",
       "sample1001         GG        TT        CC        GG        CC       CT   \n",
       "...               ...       ...       ...       ...       ...      ...   \n",
       "sample995          GG        TT        TT        GG        CC       TT   \n",
       "sample996          GG        TT        CT        GG        CC       TT   \n",
       "sample997          GG        TT        CC        GG        CC      NaN   \n",
       "sample998          GG        TT        CT        GG        CC       TT   \n",
       "sample999          GG        TT        CT        GG        CC       CT   \n",
       "\n",
       "           rs4821519 rs1003500 rs5756405  \n",
       "sample_id                                 \n",
       "sample1           GG        CC        AG  \n",
       "sample10          GG        CC        AA  \n",
       "sample100         GG        CC        AG  \n",
       "sample1000        GG        CC        AA  \n",
       "sample1001        GG        CC        AG  \n",
       "...              ...       ...       ...  \n",
       "sample995         GG        CC        AG  \n",
       "sample996         GG        CC        AA  \n",
       "sample997         GG        CC        AG  \n",
       "sample998         GG        CC        AG  \n",
       "sample999         CG        CC        GG  \n",
       "\n",
       "[5175 rows x 9 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data_catalog.save(\"control_geno_each_sample\", control_all_geno_file)\n",
    "test_data_catalog.load(\"control_geno_each_sample\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
