---

title: OVP dataset


keywords: fastai
sidebar: home_sidebar

summary: "API details."
description: "API details."
nb_path: "02_datasets_OVPDataset.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_datasets_OVPDataset.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="OVPDataSetError" class="doc_header"><code>class</code> <code>OVPDataSetError</code><a href="https://github.com/corradin-lab/corradin_ovp_utils/tree/master/ #https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L39" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>OVPDataSetError</code>() :: <code>Exception</code></p>
</blockquote>
<p>Common base class for all non-exit exceptions.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse_class" class="doc_header"><code>parse_class</code><a href="https://github.com/corradin-lab/corradin_ovp_utils/tree/master/ #https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L42" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse_class</code>(<strong><code>key</code></strong>, <strong><code>class_obj</code></strong>:<code>str</code>)</p>
</blockquote>
<p>Parse and instantiate a dataset class using the configuration provided.
Args:
    config: Data set config dictionary. It <em>must</em> contain the <code>type</code> key
        with fully qualified class name.
    load_version: Version string to be used for <code>load</code> operation if
            the data set is versioned. Has no effect on the data set
            if versioning was not enabled.
    save_version: Version string to be used for <code>save</code> operation if
        the data set is versioned. Has no effect on the data set
        if versioning was not enabled.
Raises:
    DataSetError: If the function fails to parse the configuration provided.
Returns:
    2-tuple: (Dataset class object, configuration dictionary)</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span> 
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">files</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>namespace(case=GenFileFormat(file_path=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CASE_MS_chr22.gen&#39;, file_name=&#39;test_CASE_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None, current_file_path=None), control=GenFileFormat(file_path=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CONTROL_MS_chr22.gen&#39;, file_name=&#39;test_CONTROL_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None, current_file_path=None))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file_single&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">files</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>namespace(single_file=GenFileFormat(file_path=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CASE_MS_chr22.gen&#39;, file_name=&#39;test_CASE_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None, current_file_path=None))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="OVPDataset" class="doc_header"><code>class</code> <code>OVPDataset</code><a href="https://github.com/corradin-lab/corradin_ovp_utils/tree/master/ #https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L124" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>OVPDataset</code>(<strong><code>file_type</code></strong>, <strong><code>file_format</code></strong>, <strong><code>file_path</code></strong>, <strong><code>common_folder</code></strong>=<em><code>None</code></em>, <strong><code>load_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>, <strong><code>version</code></strong>:<code>Version</code>=<em><code>None</code></em>, <strong><code>credentials</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>, <strong><code>fs_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>) :: <code>AbstractVersionedDataSet</code></p>
</blockquote>
<p><code>AbstractVersionedDataSet</code> is the base class for all versioned data set
implementations. All data sets that implement versioning should extend this
abstract class and implement the methods marked as abstract.</p>
<p>Example:
::</p>

<pre><code>&gt;&gt;&gt; from pathlib import Path, PurePosixPath
&gt;&gt;&gt; import pandas as pd
&gt;&gt;&gt; from kedro.io import AbstractVersionedDataSet
&gt;&gt;&gt;
&gt;&gt;&gt;
&gt;&gt;&gt; class MyOwnDataSet(AbstractVersionedDataSet):
&gt;&gt;&gt;     def __init__(self, filepath, version, param1, param2=True):
&gt;&gt;&gt;         super().__init__(PurePosixPath(filepath), version)
&gt;&gt;&gt;         self._param1 = param1
&gt;&gt;&gt;         self._param2 = param2
&gt;&gt;&gt;
&gt;&gt;&gt;     def _load(self) -&gt; pd.DataFrame:
&gt;&gt;&gt;         load_path = self._get_load_path()
&gt;&gt;&gt;         return pd.read_csv(load_path)
&gt;&gt;&gt;
&gt;&gt;&gt;     def _save(self, df: pd.DataFrame) -&gt; None:
&gt;&gt;&gt;         save_path = self._get_save_path()
&gt;&gt;&gt;         df.to_csv(str(save_path))
&gt;&gt;&gt;
&gt;&gt;&gt;     def _exists(self) -&gt; bool:
&gt;&gt;&gt;         path = self._get_load_path()
&gt;&gt;&gt;         return Path(path.as_posix()).exists()
&gt;&gt;&gt;
&gt;&gt;&gt;     def _describe(self):
&gt;&gt;&gt;         return dict(version=self._version, param1=self._param1, param2=self._param2)

</code></pre>
<p>Example catalog.yml specification:
::</p>

<pre><code>my_dataset:
    type: &lt;path-to-my-own-dataset&gt;.MyOwnDataSet
    filepath: data/01_raw/my_data.csv
    versioned: true
    param1: &lt;param1-value&gt; # param1 is a required argument
    # param2 will be True by default</code></pre>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_conf_test_data_catalog</span> <span class="o">=</span> <span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">test_conf_test_data_catalog</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;type&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;corradin_ovp_utils.datasets.OVPDataset.OVPDataset&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_dataset</span> <span class="o">=</span> <span class="n">OVPDataset</span><span class="p">(</span><span class="o">**</span><span class="n">test_conf_test_data_catalog</span><span class="p">)</span>
<span class="n">test_genetic_dataset</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;__main__.OVPDataset at 0x1497c8acf760&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_case_file</span> <span class="o">=</span> <span class="n">test_genetic_dataset</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">case</span>
<span class="n">test_control_file</span> <span class="o">=</span> <span class="n">test_genetic_dataset</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">control</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_case_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">ValueError</span>                                Traceback (most recent call last)
<span class="ansi-green-fg">&lt;ipython-input-47-05762c8173cf&gt;</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">----&gt; 1</span><span class="ansi-red-fg"> </span>test_case_file<span class="ansi-blue-fg">.</span>get_genotypes_df<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py</span> in <span class="ansi-cyan-fg">get_genotypes_df</span><span class="ansi-blue-fg">(self, chrom, rsid_list, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    126</span>             geno_df <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>get_rsid_df<span class="ansi-blue-fg">(</span>chrom<span class="ansi-blue-fg">=</span>chrom<span class="ansi-blue-fg">,</span> rsid_list <span class="ansi-blue-fg">=</span> rsid_list<span class="ansi-blue-fg">,</span> usecols <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>gen_info_cols<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    127</span>         <span class="ansi-green-fg">else</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 128</span><span class="ansi-red-fg">             </span>geno_df <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>load<span class="ansi-blue-fg">(</span>chrom <span class="ansi-blue-fg">=</span> chrom<span class="ansi-blue-fg">,</span> usecols <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>gen_info_cols<span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    129</span>         geno_df<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;homo_ref&#34;</span><span class="ansi-blue-fg">]</span> <span class="ansi-blue-fg">=</span> geno_df<span class="ansi-blue-fg">.</span>loc<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">:</span><span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">[</span>self<span class="ansi-blue-fg">.</span>ref_col<span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">]</span> <span class="ansi-blue-fg">*</span><span class="ansi-cyan-fg">2</span>
<span class="ansi-green-intense-fg ansi-bold">    130</span>         make_het_geno_func = lambda row: &#39;&#39;.join(sorted([row[self.ref_col],

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py</span> in <span class="ansi-cyan-fg">load</span><span class="ansi-blue-fg">(self, chrom, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    104</span>     <span class="ansi-green-fg">def</span> load<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span> chrom<span class="ansi-blue-fg">=</span><span class="ansi-green-fg">None</span><span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">    105</span>         self<span class="ansi-blue-fg">.</span>current_file_path <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>get_resolved_file_path<span class="ansi-blue-fg">(</span>chrom<span class="ansi-blue-fg">)</span> <span class="ansi-red-fg"># use another attribute so `load` and `_load_unprocessed` can share</span>
<span class="ansi-green-fg">--&gt; 106</span><span class="ansi-red-fg">         </span>df <span class="ansi-blue-fg">=</span> pd<span class="ansi-blue-fg">.</span>read_csv<span class="ansi-blue-fg">(</span>filepath_or_buffer <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>current_file_path<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>self<span class="ansi-blue-fg">.</span>load_args<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    107</span>         <span class="ansi-red-fg">#df.columns = self.column_headers</span>
<span class="ansi-green-intense-fg ansi-bold">    108</span>         <span class="ansi-green-fg">return</span> df

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py</span> in <span class="ansi-cyan-fg">load_args</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-intense-fg ansi-bold">     92</span>     <span class="ansi-green-fg">def</span> load_args<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">     93</span>         load_args <span class="ansi-blue-fg">=</span> deepcopy<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">.</span>pandas_args<span class="ansi-blue-fg">)</span>
<span class="ansi-green-fg">---&gt; 94</span><span class="ansi-red-fg">         </span>col_names <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>column_headers
<span class="ansi-green-intense-fg ansi-bold">     95</span>         load_args<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;names&#34;</span><span class="ansi-blue-fg">]</span> <span class="ansi-blue-fg">=</span> col_names
<span class="ansi-green-intense-fg ansi-bold">     96</span>         <span class="ansi-green-fg">return</span> load_args

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py</span> in <span class="ansi-cyan-fg">column_headers</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-intense-fg ansi-bold">     69</span>     <span class="ansi-green-fg">def</span> column_headers<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">     70</span>         <span class="ansi-red-fg">#read the first non-header line</span>
<span class="ansi-green-fg">---&gt; 71</span><span class="ansi-red-fg">         </span>first_line_df <span class="ansi-blue-fg">=</span> next<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">.</span>_load_unprocessed<span class="ansi-blue-fg">(</span>chunksize<span class="ansi-blue-fg">=</span><span class="ansi-cyan-fg">1</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     72</span>         generator <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>col_name_generator<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     73</span>         <span class="ansi-green-fg">return</span> <span class="ansi-blue-fg">[</span>next<span class="ansi-blue-fg">(</span>generator<span class="ansi-blue-fg">)</span> <span class="ansi-green-fg">for</span> col <span class="ansi-green-fg">in</span> first_line_df<span class="ansi-blue-fg">.</span>columns<span class="ansi-blue-fg">]</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/corradin_ovp_utils/datasets/genetic_file/Gen.py</span> in <span class="ansi-cyan-fg">_load_unprocessed</span><span class="ansi-blue-fg">(self, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">     98</span>     <span class="ansi-blue-fg">@</span>delegates<span class="ansi-blue-fg">(</span>pd<span class="ansi-blue-fg">.</span>read_csv<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     99</span>     <span class="ansi-green-fg">def</span> _load_unprocessed<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 100</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">return</span> pd<span class="ansi-blue-fg">.</span>read_csv<span class="ansi-blue-fg">(</span>filepath_or_buffer <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>current_file_path<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>self<span class="ansi-blue-fg">.</span>pandas_args<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>kwargs<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    101</span> 
<span class="ansi-green-intense-fg ansi-bold">    102</span> 

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py</span> in <span class="ansi-cyan-fg">read_csv</span><span class="ansi-blue-fg">(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)</span>
<span class="ansi-green-intense-fg ansi-bold">    603</span>     kwds<span class="ansi-blue-fg">.</span>update<span class="ansi-blue-fg">(</span>kwds_defaults<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    604</span> 
<span class="ansi-green-fg">--&gt; 605</span><span class="ansi-red-fg">     </span><span class="ansi-green-fg">return</span> _read<span class="ansi-blue-fg">(</span>filepath_or_buffer<span class="ansi-blue-fg">,</span> kwds<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    606</span> 
<span class="ansi-green-intense-fg ansi-bold">    607</span> 

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py</span> in <span class="ansi-cyan-fg">_read</span><span class="ansi-blue-fg">(filepath_or_buffer, kwds)</span>
<span class="ansi-green-intense-fg ansi-bold">    455</span> 
<span class="ansi-green-intense-fg ansi-bold">    456</span>     <span class="ansi-red-fg"># Create the parser.</span>
<span class="ansi-green-fg">--&gt; 457</span><span class="ansi-red-fg">     </span>parser <span class="ansi-blue-fg">=</span> TextFileReader<span class="ansi-blue-fg">(</span>filepath_or_buffer<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>kwds<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    458</span> 
<span class="ansi-green-intense-fg ansi-bold">    459</span>     <span class="ansi-green-fg">if</span> chunksize <span class="ansi-green-fg">or</span> iterator<span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py</span> in <span class="ansi-cyan-fg">__init__</span><span class="ansi-blue-fg">(self, f, engine, **kwds)</span>
<span class="ansi-green-intense-fg ansi-bold">    812</span>             self<span class="ansi-blue-fg">.</span>options<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;has_index_names&#34;</span><span class="ansi-blue-fg">]</span> <span class="ansi-blue-fg">=</span> kwds<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;has_index_names&#34;</span><span class="ansi-blue-fg">]</span>
<span class="ansi-green-intense-fg ansi-bold">    813</span> 
<span class="ansi-green-fg">--&gt; 814</span><span class="ansi-red-fg">         </span>self<span class="ansi-blue-fg">.</span>_engine <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>_make_engine<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">.</span>engine<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    815</span> 
<span class="ansi-green-intense-fg ansi-bold">    816</span>     <span class="ansi-green-fg">def</span> close<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py</span> in <span class="ansi-cyan-fg">_make_engine</span><span class="ansi-blue-fg">(self, engine)</span>
<span class="ansi-green-intense-fg ansi-bold">   1043</span>             )
<span class="ansi-green-intense-fg ansi-bold">   1044</span>         <span class="ansi-red-fg"># error: Too many arguments for &#34;ParserBase&#34;</span>
<span class="ansi-green-fg">-&gt; 1045</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">return</span> mapping<span class="ansi-blue-fg">[</span>engine<span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">.</span>f<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">**</span>self<span class="ansi-blue-fg">.</span>options<span class="ansi-blue-fg">)</span>  <span class="ansi-red-fg"># type: ignore[call-arg]</span>
<span class="ansi-green-intense-fg ansi-bold">   1046</span> 
<span class="ansi-green-intense-fg ansi-bold">   1047</span>     <span class="ansi-green-fg">def</span> _failover_to_python<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py</span> in <span class="ansi-cyan-fg">__init__</span><span class="ansi-blue-fg">(self, src, **kwds)</span>
<span class="ansi-green-intense-fg ansi-bold">   1860</span> 
<span class="ansi-green-intense-fg ansi-bold">   1861</span>         <span class="ansi-red-fg"># open handles</span>
<span class="ansi-green-fg">-&gt; 1862</span><span class="ansi-red-fg">         </span>self<span class="ansi-blue-fg">.</span>_open_handles<span class="ansi-blue-fg">(</span>src<span class="ansi-blue-fg">,</span> kwds<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">   1863</span>         <span class="ansi-green-fg">assert</span> self<span class="ansi-blue-fg">.</span>handles <span class="ansi-green-fg">is</span> <span class="ansi-green-fg">not</span> <span class="ansi-green-fg">None</span>
<span class="ansi-green-intense-fg ansi-bold">   1864</span>         <span class="ansi-green-fg">for</span> key <span class="ansi-green-fg">in</span> <span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#34;storage_options&#34;</span><span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">&#34;encoding&#34;</span><span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">&#34;memory_map&#34;</span><span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">&#34;compression&#34;</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/parsers.py</span> in <span class="ansi-cyan-fg">_open_handles</span><span class="ansi-blue-fg">(self, src, kwds)</span>
<span class="ansi-green-intense-fg ansi-bold">   1355</span>         Let the readers open IOHanldes after they are done <span class="ansi-green-fg">with</span> their potential raises<span class="ansi-blue-fg">.</span>
<span class="ansi-green-intense-fg ansi-bold">   1356</span>         &#34;&#34;&#34;
<span class="ansi-green-fg">-&gt; 1357</span><span class="ansi-red-fg">         self.handles = get_handle(
</span><span class="ansi-green-intense-fg ansi-bold">   1358</span>             src<span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">   1359</span>             <span class="ansi-blue-fg">&#34;r&#34;</span><span class="ansi-blue-fg">,</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/common.py</span> in <span class="ansi-cyan-fg">get_handle</span><span class="ansi-blue-fg">(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)</span>
<span class="ansi-green-intense-fg ansi-bold">    556</span> 
<span class="ansi-green-intense-fg ansi-bold">    557</span>     <span class="ansi-red-fg"># open URLs</span>
<span class="ansi-green-fg">--&gt; 558</span><span class="ansi-red-fg">     ioargs = _get_filepath_or_buffer(
</span><span class="ansi-green-intense-fg ansi-bold">    559</span>         path_or_buf<span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">    560</span>         encoding<span class="ansi-blue-fg">=</span>encoding<span class="ansi-blue-fg">,</span>

<span class="ansi-green-fg">/lab/corradin_biobank/FOR_AN/OVP/corradin_ovp_utils/.venv/lib/python3.8/site-packages/pandas/io/common.py</span> in <span class="ansi-cyan-fg">_get_filepath_or_buffer</span><span class="ansi-blue-fg">(filepath_or_buffer, encoding, compression, mode, storage_options)</span>
<span class="ansi-green-intense-fg ansi-bold">    369</span>     <span class="ansi-green-fg">if</span> <span class="ansi-green-fg">not</span> is_file_like<span class="ansi-blue-fg">(</span>filepath_or_buffer<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">    370</span>         msg <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">f&#34;Invalid file path or buffer object type: {type(filepath_or_buffer)}&#34;</span>
<span class="ansi-green-fg">--&gt; 371</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">raise</span> ValueError<span class="ansi-blue-fg">(</span>msg<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    372</span> 
<span class="ansi-green-intense-fg ansi-bold">    373</span>     return IOArgs(

<span class="ansi-red-fg">ValueError</span>: Invalid file path or buffer object type: &lt;class &#39;corradin_ovp_utils.datasets.schemas.SingleFilePathSchema&#39;&gt;</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_control_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs1014626</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5762201</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>rs1004237</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs134490</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>G</td>
      <td>C</td>
      <td>GG</td>
      <td>CG</td>
      <td>CC</td>
    </tr>
    <tr>
      <th>rs1003500</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5756405</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">case_all_geno_file</span> <span class="o">=</span> <span class="n">test_case_file</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">genetic_file</span><span class="o">.</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_case_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span>
<span class="n">case_all_geno_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">control_all_geno_file</span> <span class="o">=</span> <span class="n">test_control_file</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">genetic_file</span><span class="o">.</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_control_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span>
<span class="n">control_all_geno_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>5175 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="ow">not</span> <span class="n">case_all_geno_file</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">control_all_geno_file</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;case_geno_each_sample&quot;</span><span class="p">,</span> <span class="n">case_all_geno_file</span><span class="p">)</span>
<span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;case_geno_each_sample&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;control_geno_each_sample&quot;</span><span class="p">,</span> <span class="n">control_all_geno_file</span><span class="p">)</span>
<span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;control_geno_each_sample&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>5175 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

