---

title: module name here


keywords: fastai
sidebar: home_sidebar

summary: "API details."
description: "API details."
nb_path: "02_datasets_OVPDataset.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_datasets_OVPDataset.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="OVPDataSetError" class="doc_header"><code>class</code> <code>OVPDataSetError</code><a href="https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L37" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>OVPDataSetError</code>() :: <code>Exception</code></p>
</blockquote>
<p>Common base class for all non-exit exceptions.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse_class" class="doc_header"><code>parse_class</code><a href="https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L40" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse_class</code>(<strong><code>key</code></strong>, <strong><code>class_obj</code></strong>:<code>str</code>)</p>
</blockquote>
<p>Parse and instantiate a dataset class using the configuration provided.
Args:
    config: Data set config dictionary. It <em>must</em> contain the <code>type</code> key
        with fully qualified class name.
    load_version: Version string to be used for <code>load</code> operation if
            the data set is versioned. Has no effect on the data set
            if versioning was not enabled.
    save_version: Version string to be used for <code>save</code> operation if
        the data set is versioned. Has no effect on the data set
        if versioning was not enabled.
Raises:
    DataSetError: If the function fails to parse the configuration provided.
Returns:
    2-tuple: (Dataset class object, configuration dictionary)</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">parse_class</span><span class="p">(</span><span class="s2">&quot;file_format&quot;</span><span class="p">,</span> <span class="s2">&quot;genetic_file.GenFileFormat&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">genetic_file</span><span class="o">.</span><span class="n">GenFileFormat</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">parse_class</span><span class="p">(</span><span class="s2">&quot;file_format&quot;</span><span class="p">,</span> <span class="s2">&quot;corradin_ovp_utils.datasets.genetic_file.GenFileFormat&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">genetic_file</span><span class="o">.</span><span class="n">GenFileFormat</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">files</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>namespace(case=GenFileFormat(filepath=Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None),
          control=GenFileFormat(filepath=Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file_single&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">files</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>namespace(single_file=GenFileFormat(filepath=Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="SingleFilePathSchema" class="doc_header"><code>class</code> <code>SingleFilePathSchema</code><a href="https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L112" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>SingleFilePathSchema</code>(<strong><code>folder</code></strong>:<code>str</code>, <strong><code>full_file_name</code></strong>:<code>str</code>, <strong><code>file_name</code></strong>:<code>str</code>=<em><code>None</code></em>, <strong><code>extension</code></strong>:<code>str</code>=<em><code>None</code></em>, <strong><code>split_by_chromosome</code></strong>:<code>bool</code>=<em><code>None</code></em>) :: <code>BaseModel</code></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="MultipleFilePathSchema" class="doc_header"><code>class</code> <code>MultipleFilePathSchema</code><a href="https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L153" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>MultipleFilePathSchema</code>()</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="CaseControlFilePathSchema" class="doc_header"><code>class</code> <code>CaseControlFilePathSchema</code><a href="https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L170" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>CaseControlFilePathSchema</code>(<strong><code>case</code></strong>:<a href="/corradin_ovp_utils/datasets_OVPDataset.html#SingleFilePathSchema"><code>SingleFilePathSchema</code></a>, <strong><code>control</code></strong>:<a href="/corradin_ovp_utils/datasets_OVPDataset.html#SingleFilePathSchema"><code>SingleFilePathSchema</code></a>, <strong><code>common_folder</code></strong>:<code>Union[str, NoneType]]</code>=<em><code>None</code></em>) :: <a href="/corradin_ovp_utils/datasets_OVPDataset.html#MultipleFilePathSchema"><code>MultipleFilePathSchema</code></a></p>
</blockquote>
<p>CaseControlFilePathSchema(case: corradin_ovp_utils.datasets.OVPDataset.SingleFilePathSchema, control: corradin_ovp_utils.datasets.OVPDataset.SingleFilePathSchema, common_folder: dataclasses.InitVar[typing.Union[str, NoneType]] = None)</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<hr>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Testing-single-file-path">Testing single file path<a class="anchor-link" href="#Testing-single-file-path"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_single&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;folder&#39;: &#39;data/test_data/gen_file&#39;,
 &#39;full_file_name&#39;: &#39;test_CASE_MS_chr22.gen&#39;}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_single_file_path</span> <span class="o">=</span> <span class="n">SingleFilePathSchema</span><span class="p">(</span><span class="o">**</span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_single&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">test_genetic_file_single_file_path</span><span class="o">.</span><span class="n">full_file_path</span> <span class="o">==</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;data/test_data/gen_file/test_CASE_MS_chr22.gen&quot;</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">test_genetic_file_single_file_path</span><span class="o">.</span><span class="n">protocol</span> <span class="o">==</span> <span class="s2">&quot;file&quot;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<hr>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Testing-case-control-file-path">Testing case control file path<a class="anchor-link" href="#Testing-case-control-file-path"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: {&#39;folder&#39;: &#39;data/test_data/gen_file&#39;,
  &#39;full_file_name&#39;: &#39;test_CASE_MS_chr22.gen&#39;},
 &#39;control&#39;: {&#39;folder&#39;: &#39;data/test_data/gen_file&#39;,
  &#39;full_file_name&#39;: &#39;test_CONTROL_MS_chr22.gen&#39;}}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_cc_file_path</span> <span class="o">=</span> <span class="n">CaseControlFilePathSchema</span><span class="p">(</span><span class="o">**</span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">])</span>
<span class="n">test_genetic_file_cc_file_path</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;),
 &#39;control&#39;: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_cc_file_path</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CASE_MS_chr22.gen&#39;, file_name=&#39;test_CASE_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None),
 &#39;control&#39;: SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CONTROL_MS_chr22.gen&#39;, file_name=&#39;test_CONTROL_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Common folder</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_common_folder&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;type&#39;: &#39;nbdev_tutorial.datasets.OVPDataset.OVPDataset&#39;,
 &#39;file_format&#39;: &#39;genetic_file.GenFileFormat&#39;,
 &#39;load_args&#39;: {&#39;prob_n_cols&#39;: 3,
  &#39;initial_cols&#39;: [&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;],
  &#39;rsid_col&#39;: &#39;rsid&#39;,
  &#39;ref_col&#39;: &#39;ref&#39;,
  &#39;alt_col&#39;: &#39;alt&#39;,
  &#39;pandas_args&#39;: {&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}},
 &#39;file_type&#39;: &#39;OVPDataset.CaseControlFilePathSchema&#39;,
 &#39;file_path&#39;: {&#39;common_folder&#39;: &#39;data/test_data/gen_file&#39;,
  &#39;case&#39;: {&#39;full_file_name&#39;: &#39;test_CASE_MS_chr22.gen&#39;},
  &#39;control&#39;: {&#39;full_file_name&#39;: &#39;test_CONTROL_MS_chr22.gen&#39;}}}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_cc_file_path_common_folder</span> <span class="o">=</span> <span class="n">CaseControlFilePathSchema</span><span class="p">(</span><span class="o">**</span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_common_folder&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">])</span>
<span class="n">test_genetic_file_cc_file_path_common_folder</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;),
 &#39;control&#39;: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_cc_file_path_common_folder</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>CaseControlFilePathSchema(case=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CASE_MS_chr22.gen&#39;, file_name=&#39;test_CASE_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None), control=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CONTROL_MS_chr22.gen&#39;, file_name=&#39;test_CONTROL_MS_chr22&#39;, extension=&#39;gen&#39;, split_by_chromosome=None))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Testing-case-control-file-path-split-by-chromosome">Testing case control file path split by chromosome<a class="anchor-link" href="#Testing-case-control-file-path-split-by-chromosome"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_split_by_chrom&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;common_folder&#39;: &#39;data/test_data/gen_file&#39;,
 &#39;case&#39;: {&#39;split_by_chromosome&#39;: True,
  &#39;full_file_name&#39;: &#39;test_CASE_MS_chr{chrom_num}.gen&#39;},
 &#39;control&#39;: {&#39;split_by_chromosome&#39;: True,
  &#39;full_file_name&#39;: &#39;test_CONTROL_MS_chr{chrom_num}.gen&#39;}}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_split_by_chrom</span> <span class="o">=</span> <span class="n">CaseControlFilePathSchema</span><span class="p">(</span><span class="o">**</span><span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_split_by_chrom&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">])</span>
<span class="n">test_genetic_file_split_by_chrom</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>CaseControlFilePathSchema(case=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CASE_MS_chr{chrom_num}.gen&#39;, file_name=&#39;test_CASE_MS_chr{chrom_num}&#39;, extension=&#39;gen&#39;, split_by_chromosome=True), control=SingleFilePathSchema(folder=&#39;data/test_data/gen_file&#39;, full_file_name=&#39;test_CONTROL_MS_chr{chrom_num}.gen&#39;, file_name=&#39;test_CONTROL_MS_chr{chrom_num}&#39;, extension=&#39;gen&#39;, split_by_chromosome=True))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_split_by_chrom</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: {1: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr1.gen&#39;),
  2: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr2.gen&#39;),
  3: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr3.gen&#39;),
  4: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr4.gen&#39;),
  5: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr5.gen&#39;),
  6: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr6.gen&#39;),
  7: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr7.gen&#39;),
  8: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr8.gen&#39;),
  9: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr9.gen&#39;),
  10: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr10.gen&#39;),
  11: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr11.gen&#39;),
  12: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr12.gen&#39;),
  13: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr13.gen&#39;),
  14: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr14.gen&#39;),
  15: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr15.gen&#39;),
  16: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr16.gen&#39;),
  17: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr17.gen&#39;),
  18: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr18.gen&#39;),
  19: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr19.gen&#39;),
  20: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr20.gen&#39;),
  21: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr21.gen&#39;),
  22: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;)},
 &#39;control&#39;: {1: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr1.gen&#39;),
  2: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr2.gen&#39;),
  3: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr3.gen&#39;),
  4: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr4.gen&#39;),
  5: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr5.gen&#39;),
  6: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr6.gen&#39;),
  7: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr7.gen&#39;),
  8: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr8.gen&#39;),
  9: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr9.gen&#39;),
  10: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr10.gen&#39;),
  11: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr11.gen&#39;),
  12: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr12.gen&#39;),
  13: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr13.gen&#39;),
  14: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr14.gen&#39;),
  15: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr15.gen&#39;),
  16: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr16.gen&#39;),
  17: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr17.gen&#39;),
  18: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr18.gen&#39;),
  19: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr19.gen&#39;),
  20: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr20.gen&#39;),
  21: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr21.gen&#39;),
  22: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;)}}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<hr>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="OVPDataset" class="doc_header"><code>class</code> <code>OVPDataset</code><a href="https://github.com/hoangthienan95/corradin_ovp_utils/tree/master/corradin_ovp_utils/datasets/OVPDataset.py#L204" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>OVPDataset</code>(<strong><code>file_type</code></strong>, <strong><code>file_format</code></strong>, <strong><code>file_path</code></strong>, <strong><code>common_folder</code></strong>=<em><code>None</code></em>, <strong><code>load_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>, <strong><code>version</code></strong>:<code>Version</code>=<em><code>None</code></em>, <strong><code>credentials</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>, <strong><code>fs_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>) :: <code>AbstractVersionedDataSet</code></p>
</blockquote>
<p><code>AbstractVersionedDataSet</code> is the base class for all versioned data set
implementations. All data sets that implement versioning should extend this
abstract class and implement the methods marked as abstract.</p>
<p>Example:
::</p>

<pre><code>&gt;&gt;&gt; from pathlib import Path, PurePosixPath
&gt;&gt;&gt; import pandas as pd
&gt;&gt;&gt; from kedro.io import AbstractVersionedDataSet
&gt;&gt;&gt;
&gt;&gt;&gt;
&gt;&gt;&gt; class MyOwnDataSet(AbstractVersionedDataSet):
&gt;&gt;&gt;     def __init__(self, filepath, version, param1, param2=True):
&gt;&gt;&gt;         super().__init__(PurePosixPath(filepath), version)
&gt;&gt;&gt;         self._param1 = param1
&gt;&gt;&gt;         self._param2 = param2
&gt;&gt;&gt;
&gt;&gt;&gt;     def _load(self) -&gt; pd.DataFrame:
&gt;&gt;&gt;         load_path = self._get_load_path()
&gt;&gt;&gt;         return pd.read_csv(load_path)
&gt;&gt;&gt;
&gt;&gt;&gt;     def _save(self, df: pd.DataFrame) -&gt; None:
&gt;&gt;&gt;         save_path = self._get_save_path()
&gt;&gt;&gt;         df.to_csv(str(save_path))
&gt;&gt;&gt;
&gt;&gt;&gt;     def _exists(self) -&gt; bool:
&gt;&gt;&gt;         path = self._get_load_path()
&gt;&gt;&gt;         return Path(path.as_posix()).exists()
&gt;&gt;&gt;
&gt;&gt;&gt;     def _describe(self):
&gt;&gt;&gt;         return dict(version=self._version, param1=self._param1, param2=self._param2)

</code></pre>
<p>Example catalog.yml specification:
::</p>

<pre><code>my_dataset:
    type: &lt;path-to-my-own-dataset&gt;.MyOwnDataSet
    filepath: data/01_raw/my_data.csv
    versioned: true
    param1: &lt;param1-value&gt; # param1 is a required argument
    # param2 will be True by default</code></pre>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_conf_test_data_catalog</span> <span class="o">=</span> <span class="n">conf_test_data_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">test_conf_test_data_catalog</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;type&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;nbdev_tutorial.datasets.OVPDataset.OVPDataset&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_dataset</span> <span class="o">=</span> <span class="n">OVPDataset</span><span class="p">(</span><span class="o">**</span><span class="n">test_conf_test_data_catalog</span><span class="p">)</span>
<span class="n">test_genetic_dataset</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;__main__.OVPDataset at 0x7fb80817cfd0&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_case_file</span> <span class="o">=</span> <span class="n">test_genetic_dataset</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">case</span>
<span class="n">test_control_file</span> <span class="o">=</span> <span class="n">test_genetic_dataset</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">control</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_case_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs1014626</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5762201</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>rs1004237</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs134490</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>G</td>
      <td>C</td>
      <td>GG</td>
      <td>CG</td>
      <td>CC</td>
    </tr>
    <tr>
      <th>rs1003500</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5756405</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_control_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs1014626</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5762201</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>rs1004237</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs134490</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>G</td>
      <td>C</td>
      <td>GG</td>
      <td>CG</td>
      <td>CC</td>
    </tr>
    <tr>
      <th>rs1003500</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5756405</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">case_all_geno_file</span> <span class="o">=</span> <span class="n">test_case_file</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">genetic_file</span><span class="o">.</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_case_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span>
<span class="n">case_all_geno_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">control_all_geno_file</span> <span class="o">=</span> <span class="n">test_control_file</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">genetic_file</span><span class="o">.</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_control_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span>
<span class="n">control_all_geno_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>5175 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="ow">not</span> <span class="n">case_all_geno_file</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">control_all_geno_file</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;case_geno_each_sample&quot;</span><span class="p">,</span> <span class="n">case_all_geno_file</span><span class="p">)</span>
<span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;case_geno_each_sample&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_data_catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;control_geno_each_sample&quot;</span><span class="p">,</span> <span class="n">control_all_geno_file</span><span class="p">)</span>
<span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;control_geno_each_sample&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>5175 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

