---

title: module name here


keywords: fastai
sidebar: home_sidebar

summary: "API details."
description: "API details."
nb_path: "extract_genetic_data.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: extract_genetic_data.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_loader</span> <span class="o">=</span> <span class="n">ConfigLoader</span><span class="p">(</span><span class="s2">&quot;conf/base&quot;</span><span class="p">)</span>
<span class="n">conf_catalog</span> <span class="o">=</span> <span class="n">conf_loader</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;catalog*.yaml&quot;</span><span class="p">)</span>
<span class="n">catalog</span> <span class="o">=</span> <span class="n">DataCatalog</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf_catalog</span><span class="p">)</span>
<span class="n">gen_file_path</span> <span class="o">=</span> <span class="s2">&quot;data/test_data/gen_file/test_CASE_MS_chr22.gen&quot;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">catalog</span><span class="o">.</span><span class="n">list</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;genetic_file&#39;,
 &#39;genetic_file_common_folder&#39;,
 &#39;genetic_file_single&#39;,
 &#39;geno_each_sample&#39;,
 &#39;case_geno_each_sample&#39;,
 &#39;control_geno_each_sample&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">_file_path</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;),
 &#39;control&#39;: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_gen</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gen_file_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">,</span>
            <span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rsid&quot;</span><span class="p">],</span>
            <span class="n">header</span><span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">test_gen</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rsid</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>rs77948203</td>
    </tr>
    <tr>
      <th>1</th>
      <td>rs1014626</td>
    </tr>
    <tr>
      <th>2</th>
      <td>rs9610458</td>
    </tr>
    <tr>
      <th>3</th>
      <td>rs5762201</td>
    </tr>
    <tr>
      <th>4</th>
      <td>rs1004237</td>
    </tr>
    <tr>
      <th>5</th>
      <td>rs134490</td>
    </tr>
    <tr>
      <th>6</th>
      <td>rs4821519</td>
    </tr>
    <tr>
      <th>7</th>
      <td>rs1003500</td>
    </tr>
    <tr>
      <th>8</th>
      <td>rs5756405</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_gen</span><span class="p">[</span><span class="s2">&quot;rsid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;rs77948203&#39;,
 &#39;rs1014626&#39;,
 &#39;rs9610458&#39;,
 &#39;rs5762201&#39;,
 &#39;rs1004237&#39;,
 &#39;rs134490&#39;,
 &#39;rs4821519&#39;,
 &#39;rs1003500&#39;,
 &#39;rs5756405&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">query_rsids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs134490&quot;</span><span class="p">]</span>
<span class="n">query_rsids</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;rs77948203&#39;, &#39;rs134490&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">query_row_index_list</span> <span class="o">=</span> <span class="n">test_gen</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;rsid == @query_rsids&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">query_row_index_list</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[0, 5]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gen_file_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">,</span>
            <span class="n">skiprows</span><span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">query_row_index_list</span><span class="p">,</span>
            <span class="c1">#iterator = True,</span>
            <span class="n">chunksize</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">next</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gen_file_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">,</span>
            <span class="n">skiprows</span><span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">query_row_index_list</span><span class="p">,</span>
            <span class="c1">#iterator = True,</span>
            <span class="n">chunksize</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span> <span class="kc">None</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
      <th>9</th>
      <th>...</th>
      <th>29311</th>
      <th>29312</th>
      <th>29313</th>
      <th>29314</th>
      <th>29315</th>
      <th>29316</th>
      <th>29317</th>
      <th>29318</th>
      <th>29319</th>
      <th>29320</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>---</td>
      <td>rs77948203</td>
      <td>21249165</td>
      <td>G</td>
      <td>A</td>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1</td>
      <td>0.0</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>1 rows × 29321 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">GenFile</span><span class="p">:</span>
    <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">rsid_col</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">snp_col</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">snps</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>
    
    
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gen_file_path</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">,</span>
            <span class="n">usecols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rsid&quot;</span><span class="p">],</span>
            <span class="n">skiprows</span><span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">query_row_index_list</span><span class="p">,</span>
            <span class="n">iterator</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span> <span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">f</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="SingleFilePathSchema" class="doc_header"><code>class</code> <code>SingleFilePathSchema</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L33" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>SingleFilePathSchema</code>(<strong><code>folder</code></strong>:<code>str</code>, <strong><code>full_file_name</code></strong>:<code>str</code>, <strong><code>file_name</code></strong>:<code>str</code>=<em><code>None</code></em>, <strong><code>extension</code></strong>:<code>str</code>=<em><code>None</code></em>, <strong><code>split_by_chromosome</code></strong>:<code>bool</code>=<em><code>None</code></em>) :: <code>BaseModel</code></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="CaseControlFilePathSchema" class="doc_header"><code>class</code> <code>CaseControlFilePathSchema</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L67" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>CaseControlFilePathSchema</code>(<strong><code>case</code></strong>:<a href="/nbdev_tutorial/extract_genetic_data.html#SingleFilePathSchema"><code>SingleFilePathSchema</code></a>, <strong><code>control</code></strong>:<a href="/nbdev_tutorial/extract_genetic_data.html#SingleFilePathSchema"><code>SingleFilePathSchema</code></a>, <strong><code>common_folder</code></strong>:<code>Union[str, NoneType]]</code>=<em><code>None</code></em>)</p>
</blockquote>
<p>CaseControlFilePathSchema(case: nbdev_tutorial.datasets.genetic_datasets.SingleFilePathSchema, control: nbdev_tutorial.datasets.genetic_datasets.SingleFilePathSchema, common_folder: dataclasses.InitVar[typing.Union[str, NoneType]] = None)</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file_common_folder&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;),
 &#39;control&#39;: Path(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_single&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;folder&#39;: &#39;data/test_data/gen_file&#39;,
 &#39;full_file_name&#39;: &#39;test_CASE_MS_chr22.gen&#39;}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_single_file_path</span> <span class="o">=</span> <span class="n">SingleFilePathSchema</span><span class="p">(</span><span class="o">**</span><span class="n">conf_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file_single&quot;</span><span class="p">][</span><span class="s2">&quot;file_path&quot;</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">test_genetic_file_single_file_path</span><span class="o">.</span><span class="n">full_file_path</span> <span class="o">==</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;data/test_data/gen_file/test_CASE_MS_chr22.gen&quot;</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">test_genetic_file_single_file_path</span><span class="o">.</span><span class="n">protocol</span> <span class="o">==</span> <span class="s2">&quot;file&quot;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="MissingAttributeError" class="doc_header"><code>class</code> <code>MissingAttributeError</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L105" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>MissingAttributeError</code>() :: <code>Exception</code></p>
</blockquote>
<p>Common base class for all non-exit exceptions.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="requires" class="doc_header"><code>requires</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L108" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>requires</code>(<strong>*<code>required_attrs</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="GenFileFormat" class="doc_header"><code>class</code> <code>GenFileFormat</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L120" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>GenFileFormat</code>(<strong><code>filepath</code></strong>:<code>Union</code>[<code>str</code>, <code>Path</code>], <strong><code>prob_n_cols</code></strong>:<code>int</code>, <strong><code>initial_cols</code></strong>:<code>List</code>[<code>str</code>], <strong><code>rsid_col</code></strong>:<code>str</code>, <strong><code>ref_col</code></strong>:<code>str</code>, <strong><code>alt_col</code></strong>:<code>str</code>, <strong><code>ref_alt_delim</code></strong>:<code>str</code>=<em><code>None</code></em>, <strong><code>pandas_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>], <strong><code>sample_ids</code></strong>:<code>str</code>=<em><code>None</code></em>) :: <code>BaseModel</code></p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">conf_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">][</span><span class="s2">&quot;load_args&quot;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;prob_n_cols&#39;: 3,
 &#39;initial_cols&#39;: [&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;],
 &#39;rsid_col&#39;: &#39;rsid&#39;,
 &#39;ref_col&#39;: &#39;ref&#39;,
 &#39;alt_col&#39;: &#39;alt&#39;,
 &#39;pandas_args&#39;: {&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_file_single_file_path</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile_format</span> <span class="o">=</span> <span class="n">GenFileFormat</span><span class="p">(</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">test_genetic_file_single_file_path</span><span class="o">.</span><span class="n">full_file_path</span><span class="p">,</span>
                                    <span class="o">**</span><span class="n">conf_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">][</span><span class="s2">&quot;load_args&quot;</span><span class="p">])</span>
<span class="n">test_genfile_format</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>GenFileFormat(filepath=Path(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;), prob_n_cols=3, initial_cols=[&#39;dashes&#39;, &#39;rsid&#39;, &#39;position&#39;, &#39;ref&#39;, &#39;alt&#39;], rsid_col=&#39;rsid&#39;, ref_col=&#39;ref&#39;, alt_col=&#39;alt&#39;, ref_alt_delim=None, pandas_args={&#39;sep&#39;: &#39; &#39;, &#39;header&#39;: None}, sample_ids=None)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">load</span><span class="p">()</span> <span class="c1">#genotypes()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dashes</th>
      <th>rsid</th>
      <th>position</th>
      <th>ref</th>
      <th>alt</th>
      <th>sample1_1</th>
      <th>sample1_2</th>
      <th>sample1_3</th>
      <th>sample2_1</th>
      <th>sample2_2</th>
      <th>...</th>
      <th>sample9769_3</th>
      <th>sample9770_1</th>
      <th>sample9770_2</th>
      <th>sample9770_3</th>
      <th>sample9771_1</th>
      <th>sample9771_2</th>
      <th>sample9771_3</th>
      <th>sample9772_1</th>
      <th>sample9772_2</th>
      <th>sample9772_3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>---</td>
      <td>rs77948203</td>
      <td>21249165</td>
      <td>G</td>
      <td>A</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>---</td>
      <td>rs1014626</td>
      <td>21461017</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0.000</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>---</td>
      <td>rs9610458</td>
      <td>22205353</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1.000</td>
      <td>0.000</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>---</td>
      <td>rs5762201</td>
      <td>27888455</td>
      <td>A</td>
      <td>G</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0.012</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>---</td>
      <td>rs1004237</td>
      <td>28068501</td>
      <td>C</td>
      <td>T</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>---</td>
      <td>rs134490</td>
      <td>28730175</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.232</td>
      <td>0.768</td>
      <td>0</td>
      <td>0.014</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0.356</td>
      <td>0.644</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>6</th>
      <td>---</td>
      <td>rs4821519</td>
      <td>37102100</td>
      <td>G</td>
      <td>C</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>0</td>
      <td>1.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>---</td>
      <td>rs1003500</td>
      <td>37262769</td>
      <td>C</td>
      <td>T</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>---</td>
      <td>rs5756405</td>
      <td>37310954</td>
      <td>A</td>
      <td>G</td>
      <td>0</td>
      <td>1.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>9 rows × 29321 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_header</span> <span class="o">=</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">column_headers</span>
<span class="k">assert</span> <span class="n">test_header</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;sample9772_3&#39;</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_rsid_df</span><span class="p">([</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs9610458&quot;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dashes</th>
      <th>rsid</th>
      <th>position</th>
      <th>ref</th>
      <th>alt</th>
      <th>sample1_1</th>
      <th>sample1_2</th>
      <th>sample1_3</th>
      <th>sample2_1</th>
      <th>sample2_2</th>
      <th>...</th>
      <th>sample9769_3</th>
      <th>sample9770_1</th>
      <th>sample9770_2</th>
      <th>sample9770_3</th>
      <th>sample9771_1</th>
      <th>sample9771_2</th>
      <th>sample9771_3</th>
      <th>sample9772_1</th>
      <th>sample9772_2</th>
      <th>sample9772_3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>---</td>
      <td>rs77948203</td>
      <td>21249165</td>
      <td>G</td>
      <td>A</td>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1</td>
      <td>0.0</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>---</td>
      <td>rs9610458</td>
      <td>22205353</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>2 rows × 29321 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">next</span><span class="p">(</span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_rsid_df</span><span class="p">([</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs9610458&quot;</span><span class="p">],</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dashes</th>
      <th>rsid</th>
      <th>position</th>
      <th>ref</th>
      <th>alt</th>
      <th>sample1_1</th>
      <th>sample1_2</th>
      <th>sample1_3</th>
      <th>sample2_1</th>
      <th>sample2_2</th>
      <th>...</th>
      <th>sample9769_3</th>
      <th>sample9770_1</th>
      <th>sample9770_2</th>
      <th>sample9770_3</th>
      <th>sample9771_1</th>
      <th>sample9771_2</th>
      <th>sample9771_3</th>
      <th>sample9772_1</th>
      <th>sample9772_2</th>
      <th>sample9772_3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>---</td>
      <td>rs77948203</td>
      <td>21249165</td>
      <td>G</td>
      <td>A</td>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1</td>
      <td>0.0</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>1 rows × 29321 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs1014626</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5762201</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>rs1004237</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs134490</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>G</td>
      <td>C</td>
      <td>GG</td>
      <td>CG</td>
      <td>CC</td>
    </tr>
    <tr>
      <th>rs1003500</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5756405</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">(</span><span class="n">rsid_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs9610458&quot;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">num_samples</span> <span class="o">==</span> <span class="mi">9772</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="row_vectorize" class="doc_header"><code>row_vectorize</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L262" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>row_vectorize</code>(<strong><code>f</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="triplicate_converter" class="doc_header"><code>triplicate_converter</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L271" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>triplicate_converter</code>(<strong><code>sample_df</code></strong>, <strong><code>genotype_df</code></strong>, <strong><code>sample_id</code></strong>, <strong><code>high_lim</code></strong>=<em><code>0.9</code></em>, <strong><code>low_lim</code></strong>=<em><code>0.3</code></em>, <strong><code>NA_val</code></strong>=<em><code>'NA'</code></em>)</p>
</blockquote>
<p>takes in allele_list in gen triplicate format and
return new_list in letter format</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">first_sample</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">sample_columns_iter</span><span class="p">(</span><span class="n">rsid_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs9610458&quot;</span><span class="p">,</span> <span class="s2">&quot;rs4821519&quot;</span><span class="p">]))</span>
<span class="n">first_sample</span><span class="o">.</span><span class="n">sample_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample1_1</th>
      <th>sample1_2</th>
      <th>sample1_3</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs9610458</th>
      <td>0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>1</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">first_sample</span><span class="o">.</span><span class="n">sample_id</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;sample1&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">triplicate_converter</span><span class="p">(</span><span class="n">first_sample</span><span class="o">.</span><span class="n">sample_df</span><span class="p">,</span>
                     <span class="n">sample_id</span> <span class="o">=</span> <span class="n">first_sample</span><span class="o">.</span><span class="n">sample_id</span><span class="p">,</span>
                    <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>sample_id</th>
      <th>sample1</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs9610458</th>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile_format</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">rsid_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs9610458&quot;</span><span class="p">,</span> <span class="s2">&quot;rs4821519&quot;</span><span class="p">],</span>
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>sample_id</th>
      <th>sample1</th>
      <th>sample10</th>
      <th>sample100</th>
      <th>sample1000</th>
      <th>sample1001</th>
      <th>sample1002</th>
      <th>sample1003</th>
      <th>sample1004</th>
      <th>sample1005</th>
      <th>sample1006</th>
      <th>...</th>
      <th>sample990</th>
      <th>sample991</th>
      <th>sample992</th>
      <th>sample993</th>
      <th>sample994</th>
      <th>sample995</th>
      <th>sample996</th>
      <th>sample997</th>
      <th>sample998</th>
      <th>sample999</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs9610458</th>
      <td>TT</td>
      <td>TT</td>
      <td>TT</td>
      <td>CT</td>
      <td>TT</td>
      <td>CC</td>
      <td>CT</td>
      <td>CT</td>
      <td>CT</td>
      <td>CC</td>
      <td>...</td>
      <td>TT</td>
      <td>CT</td>
      <td>CT</td>
      <td>CT</td>
      <td>TT</td>
      <td>CT</td>
      <td>CT</td>
      <td>CT</td>
      <td>CT</td>
      <td>CT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>CG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>...</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>CC</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>2 rows × 9772 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>If you don't input a list of rsids, it will use all the rsids in the dataset, which could take a <strong>LONG TIME</strong></p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_geno_each_sample</span> <span class="o">=</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span> <span class="c1"># transposed</span>

<span class="n">test_geno_each_sample</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;geno_each_sample&quot;</span><span class="p">,</span> <span class="n">test_geno_each_sample</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_geno_each_sample</span><span class="p">[[</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs1014626&quot;</span><span class="p">,</span> <span class="s2">&quot;rs1004237&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs1014626&quot;</span><span class="p">,</span> <span class="s2">&quot;rs1004237&quot;</span><span class="p">])[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span> <span class="c1">#.melt(value_vars = [&quot;rs77948203&quot;, &quot;rs1014626&quot;], ignore_index=False).reset_index().groupby([&quot;rsid&quot;, &quot;value&quot;])[&quot;index&quot;].unique()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs1004237</th>
      <th>sample_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AA</td>
      <td>TT</td>
      <td>CC</td>
      <td>[sample1445, sample1545, sample2271, sample237...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AA</td>
      <td>TT</td>
      <td>NA</td>
      <td>[sample465, sample6764]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AG</td>
      <td>CT</td>
      <td>CC</td>
      <td>[sample5663, sample5840]</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AG</td>
      <td>NA</td>
      <td>CC</td>
      <td>[sample297]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AG</td>
      <td>TT</td>
      <td>CC</td>
      <td>[sample1000, sample1002, sample1025, sample102...</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AG</td>
      <td>TT</td>
      <td>NA</td>
      <td>[sample6578, sample678]</td>
    </tr>
    <tr>
      <th>6</th>
      <td>GG</td>
      <td>CT</td>
      <td>CC</td>
      <td>[sample1264, sample1359, sample1960, sample220...</td>
    </tr>
    <tr>
      <th>7</th>
      <td>GG</td>
      <td>NA</td>
      <td>CC</td>
      <td>[sample1378, sample1496, sample2247, sample391...</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>[sample1, sample10, sample100, sample1001, sam...</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GG</td>
      <td>TT</td>
      <td>NA</td>
      <td>[sample1601, sample2147, sample4016, sample421...</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_geno_each_sample</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genfile</span> <span class="o">=</span> <span class="n">test_genfile_format</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="n">test_genfile</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dashes</th>
      <th>rsid</th>
      <th>position</th>
      <th>ref</th>
      <th>alt</th>
      <th>sample1_1</th>
      <th>sample1_2</th>
      <th>sample1_3</th>
      <th>sample2_1</th>
      <th>sample2_2</th>
      <th>...</th>
      <th>sample9769_3</th>
      <th>sample9770_1</th>
      <th>sample9770_2</th>
      <th>sample9770_3</th>
      <th>sample9771_1</th>
      <th>sample9771_2</th>
      <th>sample9771_3</th>
      <th>sample9772_1</th>
      <th>sample9772_2</th>
      <th>sample9772_3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>---</td>
      <td>rs77948203</td>
      <td>21249165</td>
      <td>G</td>
      <td>A</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>---</td>
      <td>rs1014626</td>
      <td>21461017</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0.000</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>---</td>
      <td>rs9610458</td>
      <td>22205353</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1.000</td>
      <td>0.000</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>---</td>
      <td>rs5762201</td>
      <td>27888455</td>
      <td>A</td>
      <td>G</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0.012</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0.000</td>
      <td>1.000</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>---</td>
      <td>rs1004237</td>
      <td>28068501</td>
      <td>C</td>
      <td>T</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>---</td>
      <td>rs134490</td>
      <td>28730175</td>
      <td>C</td>
      <td>T</td>
      <td>0</td>
      <td>0.232</td>
      <td>0.768</td>
      <td>0</td>
      <td>0.014</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0.356</td>
      <td>0.644</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>6</th>
      <td>---</td>
      <td>rs4821519</td>
      <td>37102100</td>
      <td>G</td>
      <td>C</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>0</td>
      <td>1.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>---</td>
      <td>rs1003500</td>
      <td>37262769</td>
      <td>C</td>
      <td>T</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>---</td>
      <td>rs5756405</td>
      <td>37310954</td>
      <td>A</td>
      <td>G</td>
      <td>0</td>
      <td>1.000</td>
      <td>0.000</td>
      <td>1</td>
      <td>0.000</td>
      <td>...</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>9 rows × 29321 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="FILE_FORMAT_ENUM" class="doc_header"><code>FILE_FORMAT_ENUM</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L292" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Enum</code> = [GenFile]</p>
</blockquote>
<p>An enumeration.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="FILE_TYPE_ENUM" class="doc_header"><code>FILE_TYPE_ENUM</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L295" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Enum</code> = [CC, S]</p>
</blockquote>
<p>An enumeration.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="OVPDataset" class="doc_header"><code>class</code> <code>OVPDataset</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L302" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>OVPDataset</code>(<strong><code>file_type</code></strong>, <strong><code>file_format</code></strong>, <strong><code>file_path</code></strong>, <strong><code>common_folder</code></strong>=<em><code>None</code></em>, <strong><code>load_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>, <strong><code>version</code></strong>:<code>Version</code>=<em><code>None</code></em>, <strong><code>credentials</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>, <strong><code>fs_args</code></strong>:<code>Dict</code>[<code>str</code>, <code>Any</code>]=<em><code>None</code></em>) :: <code>AbstractVersionedDataSet</code></p>
</blockquote>
<p><code>AbstractVersionedDataSet</code> is the base class for all versioned data set
implementations. All data sets that implement versioning should extend this
abstract class and implement the methods marked as abstract.</p>
<p>Example:
::</p>

<pre><code>&gt;&gt;&gt; from pathlib import Path, PurePosixPath
&gt;&gt;&gt; import pandas as pd
&gt;&gt;&gt; from kedro.io import AbstractVersionedDataSet
&gt;&gt;&gt;
&gt;&gt;&gt;
&gt;&gt;&gt; class MyOwnDataSet(AbstractVersionedDataSet):
&gt;&gt;&gt;     def __init__(self, filepath, version, param1, param2=True):
&gt;&gt;&gt;         super().__init__(PurePosixPath(filepath), version)
&gt;&gt;&gt;         self._param1 = param1
&gt;&gt;&gt;         self._param2 = param2
&gt;&gt;&gt;
&gt;&gt;&gt;     def _load(self) -&gt; pd.DataFrame:
&gt;&gt;&gt;         load_path = self._get_load_path()
&gt;&gt;&gt;         return pd.read_csv(load_path)
&gt;&gt;&gt;
&gt;&gt;&gt;     def _save(self, df: pd.DataFrame) -&gt; None:
&gt;&gt;&gt;         save_path = self._get_save_path()
&gt;&gt;&gt;         df.to_csv(str(save_path))
&gt;&gt;&gt;
&gt;&gt;&gt;     def _exists(self) -&gt; bool:
&gt;&gt;&gt;         path = self._get_load_path()
&gt;&gt;&gt;         return Path(path.as_posix()).exists()
&gt;&gt;&gt;
&gt;&gt;&gt;     def _describe(self):
&gt;&gt;&gt;         return dict(version=self._version, param1=self._param1, param2=self._param2)

</code></pre>
<p>Example catalog.yml specification:
::</p>

<pre><code>my_dataset:
    type: &lt;path-to-my-own-dataset&gt;.MyOwnDataSet
    filepath: data/01_raw/my_data.csv
    versioned: true
    param1: &lt;param1-value&gt; # param1 is a required argument
    # param2 will be True by default</code></pre>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_conf_catalog</span> <span class="o">=</span> <span class="n">conf_catalog</span><span class="p">[</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">test_conf_catalog</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;type&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;nbdev_tutorial.datasets.genetic_datasets.GenFileDataset&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_genetic_dataset</span> <span class="o">=</span> <span class="n">OVPDataset</span><span class="p">(</span><span class="o">**</span><span class="n">test_conf_catalog</span><span class="p">)</span>
<span class="n">test_genetic_dataset</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;__main__.OVPDataset at 0x7fc00898d910&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_case_file</span> <span class="o">=</span> <span class="n">test_genetic_dataset</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">case</span>
<span class="n">test_control_file</span> <span class="o">=</span> <span class="n">test_genetic_dataset</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">control</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_case_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs1014626</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5762201</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>rs1004237</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs134490</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>G</td>
      <td>C</td>
      <td>GG</td>
      <td>CG</td>
      <td>CC</td>
    </tr>
    <tr>
      <th>rs1003500</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5756405</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">test_control_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ref</th>
      <th>alt</th>
      <th>homo_ref</th>
      <th>het</th>
      <th>homo_alt</th>
    </tr>
    <tr>
      <th>rsid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>rs77948203</th>
      <td>G</td>
      <td>A</td>
      <td>GG</td>
      <td>AG</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>rs1014626</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs9610458</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5762201</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>rs1004237</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs134490</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs4821519</th>
      <td>G</td>
      <td>C</td>
      <td>GG</td>
      <td>CG</td>
      <td>CC</td>
    </tr>
    <tr>
      <th>rs1003500</th>
      <td>C</td>
      <td>T</td>
      <td>CC</td>
      <td>CT</td>
      <td>TT</td>
    </tr>
    <tr>
      <th>rs5756405</th>
      <td>A</td>
      <td>G</td>
      <td>AA</td>
      <td>AG</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">case_all_geno_file</span> <span class="o">=</span> <span class="n">test_case_file</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_case_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span>
<span class="n">case_all_geno_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">control_all_geno_file</span> <span class="o">=</span> <span class="n">test_control_file</span><span class="o">.</span><span class="n">apply_func_to_all_samples</span><span class="p">(</span><span class="n">triplicate_converter</span><span class="p">,</span> 
                                              <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">test_control_file</span><span class="o">.</span><span class="n">get_genotypes_df</span><span class="p">())</span><span class="o">.</span><span class="n">T</span>
<span class="n">control_all_geno_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>5175 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="ow">not</span> <span class="n">case_all_geno_file</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">control_all_geno_file</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;case_geno_each_sample&quot;</span><span class="p">,</span> <span class="n">case_all_geno_file</span><span class="p">)</span>
<span class="n">catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;case_geno_each_sample&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CC</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
  </tbody>
</table>
<p>9772 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">catalog</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;control_geno_each_sample&quot;</span><span class="p">,</span> <span class="n">control_all_geno_file</span><span class="p">)</span>
<span class="n">catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;control_geno_each_sample&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/nbdev_tutorial_2nd_try/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs1014626</th>
      <th>rs9610458</th>
      <th>rs5762201</th>
      <th>rs1004237</th>
      <th>rs134490</th>
      <th>rs4821519</th>
      <th>rs1003500</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sample1</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample10</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample100</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample1000</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample1001</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sample995</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample996</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>sample997</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>GG</td>
      <td>CC</td>
      <td>NaN</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample998</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>GG</td>
      <td>CC</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>sample999</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>CG</td>
      <td>CC</td>
      <td>GG</td>
    </tr>
  </tbody>
</table>
<p>5175 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="say_hello" class="doc_header"><code>say_hello</code><a href="https://github.com/hoangthienan95/nbdev_tutorial/tree/master/nbdev_tutorial/datasets/genetic_datasets.py#L445" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>say_hello</code>(<strong><code>to</code></strong>)</p>
</blockquote>
<p>Say hello to somebody</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">say_hello</span><span class="p">(</span><span class="s2">&quot;Sylvain&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;Hello Sylvain!&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">display</span><span class="p">,</span><span class="n">SVG</span>
<span class="n">display</span><span class="p">(</span><span class="n">SVG</span><span class="p">(</span><span class="s1">&#39;&lt;svg height=&quot;100&quot;&gt;&lt;circle cx=&quot;50&quot; cy=&quot;50&quot; r=&quot;40&quot;/&gt;&lt;/svg&gt;&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_svg output_subarea ">
<svg height="100"><circle cx="50" cy="50" r="40"/></svg>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">say_hello</span><span class="p">(</span><span class="s2">&quot;Jeremy&quot;</span><span class="p">)</span><span class="o">==</span><span class="s2">&quot;Hello Jeremy!&quot;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

