---

title: Outside Variants utilities functions


keywords: fastai
sidebar: home_sidebar

summary: "Library of functions for the Outside Variant Pipeline to create unifying API and facilitate interactive exploration, so anyone can run the Outside Variant Pipeline"
description: "Library of functions for the Outside Variant Pipeline to create unifying API and facilitate interactive exploration, so anyone can run the Outside Variant Pipeline"
nb_path: "index.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: index.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="https://github.com/corradin-lab/corradin-ovp-utils/actions/workflows/main.yml"><img src="https://github.com/corradin-lab/corradin-ovp-utils/actions/workflows/main.yml/badge.svg" alt="CI"></a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Install">Install<a class="anchor-link" href="#Install"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><code>pip install corradin_ovp_utils</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="How-to-use">How to use<a class="anchor-link" href="#How-to-use"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This library is created to create a unifying API to read in different genetic data formats and specification, combine with phenotype data to create the core data structure of the outside variant pipeline. The library decouples data ingestion from downstream analyses, and make it extremely easy to extend to other input data formats while maintaining the core analysis features</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Quick-Start">Quick Start<a class="anchor-link" href="#Quick-Start"> </a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>After specifying the input in a <code>yaml</code> file (see <code>conf/base/catalog_input/genetic_file.yaml</code> and <code>conf/base/catalog_input/sample_file.yaml</code>), we can load the datasets in like this:</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">kedro.config</span> <span class="kn">import</span> <span class="n">ConfigLoader</span>
<span class="kn">from</span> <span class="nn">kedro.io</span> <span class="kn">import</span> <span class="n">DataCatalog</span>
<span class="n">conf_loader</span> <span class="o">=</span> <span class="n">ConfigLoader</span><span class="p">(</span><span class="s2">&quot;conf/base&quot;</span><span class="p">)</span>
<span class="n">conf_test_data_catalog</span> <span class="o">=</span> <span class="n">conf_loader</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;catalog*.yaml&quot;</span><span class="p">,</span> <span class="s2">&quot;catalog*/*.yaml&quot;</span><span class="p">)</span>
<span class="n">test_data_catalog</span> <span class="o">=</span> <span class="n">DataCatalog</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf_test_data_catalog</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Each of them is an <a href="/corradin_ovp_utils/datasets_OVPDataset.html#OVPDataset"><code>OVPDataset</code></a> with a unifying API</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">genetic_file</span> <span class="o">=</span> <span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;genetic_file&quot;</span><span class="p">)</span>
<span class="n">sample_file</span> <span class="o">=</span> <span class="n">test_data_catalog</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;sample_file&quot;</span><span class="p">)</span>

<span class="n">genetic_file</span><span class="p">,</span> <span class="n">sample_file</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(&lt;corradin_ovp_utils.datasets.OVPDataset.OVPDataset at 0x7fd8383aa4f0&gt;,
 &lt;corradin_ovp_utils.datasets.OVPDataset.OVPDataset at 0x7fd878c9fcd0&gt;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We can see that the <code>genetic_file</code> and <code>sample_file</code> contains two files, case and control</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">genetic_file</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: PosixPath(&#39;data/test_data/gen_file/test_CASE_MS_chr22.gen&#39;),
 &#39;control&#39;: PosixPath(&#39;data/test_data/gen_file/test_CONTROL_MS_chr22.gen&#39;)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sample_file</span><span class="o">.</span><span class="n">full_file_path</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;case&#39;: PosixPath(&#39;data/test_data/sample_file/MS_impute2_ALL_sample_out.tsv&#39;),
 &#39;control&#39;: PosixPath(&#39;data/test_data/sample_file/ALL_controls_58C_NBS_WTC2_impute2_sample_out.tsv&#39;)}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To extract the core data structure of the outside variant pipeline, you need:</p>
<ul>
<li>Genetic file</li>
<li>Sample file</li>
<li>List of Rsid to extract information</li>
</ul>
<p>Just feed these inputs into the <a href="/corradin_ovp_utils/datasets_CombinedGenoPheno.html#CombinedGenoPheno"><code>CombinedGenoPheno</code></a> object to get back a dataframe of genotypes for those SNPs for all of the samples (both case and control). <strong>This is the core data structure of the pipeline</strong>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">all_samples_geno_df</span> <span class="o">=</span> <span class="n">CombinedGenoPheno</span><span class="o">.</span><span class="n">init_from_OVPDataset</span><span class="p">(</span><span class="n">genetic_file</span><span class="p">,</span> <span class="n">sample_file</span><span class="p">,</span> <span class="n">rsid_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs9610458&quot;</span><span class="p">,</span> <span class="s2">&quot;rs134490&quot;</span><span class="p">,</span> <span class="s2">&quot;rs5756405&quot;</span><span class="p">])</span>
<span class="n">all_samples_geno_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/corradin_ovp_utils/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rsid</th>
      <th>rs77948203</th>
      <th>rs9610458</th>
      <th>rs134490</th>
      <th>rs5756405</th>
    </tr>
    <tr>
      <th>sample_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>WTCCCT473540</th>
      <td>GG</td>
      <td>TT</td>
      <td>NA</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>WTCCCT473530</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>WTCCCT473555</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>WTCCCT473426</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>WTCCCT473489</th>
      <td>GG</td>
      <td>CT</td>
      <td>NA</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>WS574632</th>
      <td>GG</td>
      <td>CT</td>
      <td>TT</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>WS574661</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>AA</td>
    </tr>
    <tr>
      <th>BLOOD294452</th>
      <td>GG</td>
      <td>CT</td>
      <td>TT</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>WTCCCT511021</th>
      <td>GG</td>
      <td>CT</td>
      <td>TT</td>
      <td>AG</td>
    </tr>
    <tr>
      <th>WTCCCT510948</th>
      <td>GG</td>
      <td>CT</td>
      <td>TT</td>
      <td>AA</td>
    </tr>
  </tbody>
</table>
<p>14947 rows × 4 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We can then take this output dataframe and do downstream analysis with it, using the functions in this library. For example, let's see the break down of genotypes grouped by the two SNPs <code>"rs77948203"</code> and <code>"rs9610458"</code></p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_geno_combination_df</span><span class="p">(</span><span class="n">geno_each_sample_df</span><span class="o">=</span><span class="n">all_samples_geno_df</span><span class="p">,</span> 
                       <span class="n">rsid_list</span><span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs9610458&quot;</span><span class="p">],</span> <span class="n">as_df</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/corradin_ovp_utils/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs9610458</th>
      <th>unique_samples_id</th>
      <th>unique_samples_count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AA</td>
      <td>CC</td>
      <td>[WTCCCT470057, WTCCCT489315, WTCCCT508408, WTC...</td>
      <td>19</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AA</td>
      <td>CT</td>
      <td>[WTCCCT474394, WTCCCT470264, WTCCCT470548, WTC...</td>
      <td>34</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AA</td>
      <td>NA</td>
      <td>[WTCCCT474448, WTCCCT508352]</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AA</td>
      <td>TT</td>
      <td>[WTCCCT474560, WTCCCT469955, WTCCCT470219, WTC...</td>
      <td>23</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AG</td>
      <td>CC</td>
      <td>[WTCCCT466268, WTCCCT489637, WTCCCT488814, WTC...</td>
      <td>360</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AG</td>
      <td>CT</td>
      <td>[WTCCCT473524, WTCCCT473551, WTCCCT489609, WTC...</td>
      <td>949</td>
    </tr>
    <tr>
      <th>6</th>
      <td>AG</td>
      <td>NA</td>
      <td>[WTCCCT489613, WTCCCT497565, WTCCCT468278, WTC...</td>
      <td>61</td>
    </tr>
    <tr>
      <th>7</th>
      <td>AG</td>
      <td>TT</td>
      <td>[WTCCCT473522, WTCCCT473497, WTCCCT473514, WTC...</td>
      <td>575</td>
    </tr>
    <tr>
      <th>8</th>
      <td>GG</td>
      <td>CC</td>
      <td>[WTCCCT473500, WTCCCT473552, WTCCCT473505, WTC...</td>
      <td>2593</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GG</td>
      <td>CT</td>
      <td>[WTCCCT473489, WTCCCT473456, WTCCCT473515, WTC...</td>
      <td>6126</td>
    </tr>
    <tr>
      <th>10</th>
      <td>GG</td>
      <td>NA</td>
      <td>[WTCCCT473549, WTCCCT489615, WTCCCT489614, WTC...</td>
      <td>423</td>
    </tr>
    <tr>
      <th>11</th>
      <td>GG</td>
      <td>TT</td>
      <td>[WTCCCT473540, WTCCCT473530, WTCCCT473555, WTC...</td>
      <td>3782</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Let's add the third SNP, <code>"rs134490"</code>, and see how the break down changes</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">triple_SNPs</span> <span class="o">=</span> <span class="n">get_geno_combination_df</span><span class="p">(</span><span class="n">geno_each_sample_df</span><span class="o">=</span><span class="n">all_samples_geno_df</span><span class="p">,</span> 
                       <span class="n">rsid_list</span><span class="o">=</span> <span class="p">[</span><span class="s2">&quot;rs77948203&quot;</span><span class="p">,</span> <span class="s2">&quot;rs9610458&quot;</span><span class="p">,</span> <span class="s2">&quot;rs134490&quot;</span><span class="p">])</span>
<span class="n">triple_SNPs</span><span class="o">.</span><span class="n">df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/corradin_ovp_utils/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs9610458</th>
      <th>rs134490</th>
      <th>unique_samples_id</th>
      <th>unique_samples_count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AA</td>
      <td>CC</td>
      <td>CT</td>
      <td>[WTCCCT508925, CCC2_MS656176, WTCCCT444162, WT...</td>
      <td>4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AA</td>
      <td>CC</td>
      <td>NA</td>
      <td>[WTCCCT490220, BLOOD293205]</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AA</td>
      <td>CC</td>
      <td>TT</td>
      <td>[WTCCCT470057, WTCCCT489315, WTCCCT508408, WTC...</td>
      <td>13</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AA</td>
      <td>CT</td>
      <td>CT</td>
      <td>[WTCCCT466178, WTCCCT468665, WTCCCT471002, WTC...</td>
      <td>8</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AA</td>
      <td>CT</td>
      <td>NA</td>
      <td>[WTCCCT474394, WTCCCT470548, WTCCCT443601]</td>
      <td>3</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AA</td>
      <td>CT</td>
      <td>TT</td>
      <td>[WTCCCT470264, WTCCCT449002, WTCCCT467316, WTC...</td>
      <td>23</td>
    </tr>
    <tr>
      <th>6</th>
      <td>AA</td>
      <td>NA</td>
      <td>CT</td>
      <td>[WTCCCT474448]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>7</th>
      <td>AA</td>
      <td>NA</td>
      <td>NA</td>
      <td>[WTCCCT508352]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>8</th>
      <td>AA</td>
      <td>TT</td>
      <td>CC</td>
      <td>[BLOOD293241]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>9</th>
      <td>AA</td>
      <td>TT</td>
      <td>CT</td>
      <td>[WTCCCT470219, WTCCCT466993, WTCCCT508309, WTC...</td>
      <td>6</td>
    </tr>
    <tr>
      <th>10</th>
      <td>AA</td>
      <td>TT</td>
      <td>NA</td>
      <td>[WTCCCT442733, WTCCCT543105, WTCCCT543369]</td>
      <td>3</td>
    </tr>
    <tr>
      <th>11</th>
      <td>AA</td>
      <td>TT</td>
      <td>TT</td>
      <td>[WTCCCT474560, WTCCCT469955, WTCCCT471095, WTC...</td>
      <td>13</td>
    </tr>
    <tr>
      <th>12</th>
      <td>AG</td>
      <td>CC</td>
      <td>CC</td>
      <td>[WTCCCT470291, WTCCCT506356, WTCCCT515627, WTC...</td>
      <td>8</td>
    </tr>
    <tr>
      <th>13</th>
      <td>AG</td>
      <td>CC</td>
      <td>CT</td>
      <td>[WTCCCT489637, WTCCCT507947, WTCCCT474442, WTC...</td>
      <td>85</td>
    </tr>
    <tr>
      <th>14</th>
      <td>AG</td>
      <td>CC</td>
      <td>NA</td>
      <td>[WTCCCT488814, WTCCCT497627, WTCCCT508819, WTC...</td>
      <td>51</td>
    </tr>
    <tr>
      <th>15</th>
      <td>AG</td>
      <td>CC</td>
      <td>TT</td>
      <td>[WTCCCT466268, WTCCCT467701, WTCCCT488998, WTC...</td>
      <td>216</td>
    </tr>
    <tr>
      <th>16</th>
      <td>AG</td>
      <td>CT</td>
      <td>CC</td>
      <td>[WTCCCT497897, WTCCCT467963, WTCCCT506246, WTC...</td>
      <td>26</td>
    </tr>
    <tr>
      <th>17</th>
      <td>AG</td>
      <td>CT</td>
      <td>CT</td>
      <td>[WTCCCT489588, WTCCCT467776, WTCCCT467779, WTC...</td>
      <td>242</td>
    </tr>
    <tr>
      <th>18</th>
      <td>AG</td>
      <td>CT</td>
      <td>NA</td>
      <td>[WTCCCT473524, WTCCCT467504, WTCCCT474443, WTC...</td>
      <td>111</td>
    </tr>
    <tr>
      <th>19</th>
      <td>AG</td>
      <td>CT</td>
      <td>TT</td>
      <td>[WTCCCT473551, WTCCCT489609, WTCCCT489577, WTC...</td>
      <td>570</td>
    </tr>
    <tr>
      <th>20</th>
      <td>AG</td>
      <td>NA</td>
      <td>CC</td>
      <td>[WTCCCT443738]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>21</th>
      <td>AG</td>
      <td>NA</td>
      <td>CT</td>
      <td>[WTCCCT468278, WTCCCT474556, WTCCCT515401, WTC...</td>
      <td>22</td>
    </tr>
    <tr>
      <th>22</th>
      <td>AG</td>
      <td>NA</td>
      <td>NA</td>
      <td>[WTCCCT489613, WTCCCT497565, WTCCCT508768, WTC...</td>
      <td>7</td>
    </tr>
    <tr>
      <th>23</th>
      <td>AG</td>
      <td>NA</td>
      <td>TT</td>
      <td>[WTCCCT470293, WTCCCT470261, WTCCCT508484, WTC...</td>
      <td>31</td>
    </tr>
    <tr>
      <th>24</th>
      <td>AG</td>
      <td>TT</td>
      <td>CC</td>
      <td>[WTCCCT507975, WTCCCT467005, WTCCCT466412, WTC...</td>
      <td>10</td>
    </tr>
    <tr>
      <th>25</th>
      <td>AG</td>
      <td>TT</td>
      <td>CT</td>
      <td>[WTCCCT489586, WTCCCT507962, WTCCCT473262, WTC...</td>
      <td>144</td>
    </tr>
    <tr>
      <th>26</th>
      <td>AG</td>
      <td>TT</td>
      <td>NA</td>
      <td>[WTCCCT473497, WTCCCT467773, WTCCCT473218, WTC...</td>
      <td>79</td>
    </tr>
    <tr>
      <th>27</th>
      <td>AG</td>
      <td>TT</td>
      <td>TT</td>
      <td>[WTCCCT473522, WTCCCT473514, WTCCCT489575, WTC...</td>
      <td>342</td>
    </tr>
    <tr>
      <th>28</th>
      <td>GG</td>
      <td>CC</td>
      <td>CC</td>
      <td>[WTCCCT489620, WTCCCT489645, WTCCCT473287, WTC...</td>
      <td>68</td>
    </tr>
    <tr>
      <th>29</th>
      <td>GG</td>
      <td>CC</td>
      <td>CT</td>
      <td>[WTCCCT473552, WTCCCT473505, WTCCCT489578, WTC...</td>
      <td>635</td>
    </tr>
    <tr>
      <th>30</th>
      <td>GG</td>
      <td>CC</td>
      <td>NA</td>
      <td>[WTCCCT489646, WTCCCT489580, WTCCCT488843, WTC...</td>
      <td>328</td>
    </tr>
    <tr>
      <th>31</th>
      <td>GG</td>
      <td>CC</td>
      <td>TT</td>
      <td>[WTCCCT473500, WTCCCT473539, WTCCCT473521, WTC...</td>
      <td>1562</td>
    </tr>
    <tr>
      <th>32</th>
      <td>GG</td>
      <td>CT</td>
      <td>CC</td>
      <td>[WTCCCT473297, WTCCCT473230, WTCCCT473244, WTC...</td>
      <td>165</td>
    </tr>
    <tr>
      <th>33</th>
      <td>GG</td>
      <td>CT</td>
      <td>CT</td>
      <td>[WTCCCT473447, WTCCCT473466, WTCCCT473492, WTC...</td>
      <td>1483</td>
    </tr>
    <tr>
      <th>34</th>
      <td>GG</td>
      <td>CT</td>
      <td>NA</td>
      <td>[WTCCCT473489, WTCCCT473499, WTCCCT473529, WTC...</td>
      <td>845</td>
    </tr>
    <tr>
      <th>35</th>
      <td>GG</td>
      <td>CT</td>
      <td>TT</td>
      <td>[WTCCCT473456, WTCCCT473515, WTCCCT473508, WTC...</td>
      <td>3633</td>
    </tr>
    <tr>
      <th>36</th>
      <td>GG</td>
      <td>NA</td>
      <td>CC</td>
      <td>[WTCCCT473436, WTCCCT469571, WTCCCT442728, WTC...</td>
      <td>4</td>
    </tr>
    <tr>
      <th>37</th>
      <td>GG</td>
      <td>NA</td>
      <td>CT</td>
      <td>[WTCCCT488883, WTCCCT474387, WTCCCT470552, WTC...</td>
      <td>88</td>
    </tr>
    <tr>
      <th>38</th>
      <td>GG</td>
      <td>NA</td>
      <td>NA</td>
      <td>[WTCCCT474593, WTCCCT469989, WTCCCT515392, WTC...</td>
      <td>61</td>
    </tr>
    <tr>
      <th>39</th>
      <td>GG</td>
      <td>NA</td>
      <td>TT</td>
      <td>[WTCCCT473549, WTCCCT489615, WTCCCT489614, WTC...</td>
      <td>270</td>
    </tr>
    <tr>
      <th>40</th>
      <td>GG</td>
      <td>TT</td>
      <td>CC</td>
      <td>[WTCCCT489604, WTCCCT467785, WTCCCT489030, WTC...</td>
      <td>91</td>
    </tr>
    <tr>
      <th>41</th>
      <td>GG</td>
      <td>TT</td>
      <td>CT</td>
      <td>[WTCCCT473454, WTCCCT489616, WTCCCT489608, WTC...</td>
      <td>872</td>
    </tr>
    <tr>
      <th>42</th>
      <td>GG</td>
      <td>TT</td>
      <td>NA</td>
      <td>[WTCCCT473540, WTCCCT473520, WTCCCT473486, WTC...</td>
      <td>552</td>
    </tr>
    <tr>
      <th>43</th>
      <td>GG</td>
      <td>TT</td>
      <td>TT</td>
      <td>[WTCCCT473530, WTCCCT473555, WTCCCT473426, WTC...</td>
      <td>2267</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We can compute basic information about these 3 SNPs</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;how many samples have at least one low quality (`NA`) genotype?&quot;</span><span class="p">,</span> <span class="n">triple_SNPs</span><span class="o">.</span><span class="n">num_samples_NA</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;how many samples have genotypes of high quality for all 3 SNPs?&quot;</span><span class="p">,</span><span class="n">triple_SNPs</span><span class="o">.</span><span class="n">total_samples_no_NA</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>how many samples have at least one low quality (`NA`) genotype? 2460
how many samples have genotypes of high quality for all 3 SNPs? 12487
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/ahoang/Documents/Learning/corradin_ovp_utils/.venv/lib/python3.8/site-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We can query based on genotype of each SNP:</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">triple_SNPs</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">rs77948203</span><span class="o">=</span> <span class="s2">&quot;AA&quot;</span><span class="p">,</span> <span class="n">rs9610458</span> <span class="o">=</span> <span class="s2">&quot;CT&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>rs77948203</th>
      <th>rs9610458</th>
      <th>rs134490</th>
      <th>unique_samples_id</th>
      <th>unique_samples_count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>AA</td>
      <td>CT</td>
      <td>CT</td>
      <td>[WTCCCT466178, WTCCCT468665, WTCCCT471002, WTC...</td>
      <td>8</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AA</td>
      <td>CT</td>
      <td>NA</td>
      <td>[WTCCCT474394, WTCCCT470548, WTCCCT443601]</td>
      <td>3</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AA</td>
      <td>CT</td>
      <td>TT</td>
      <td>[WTCCCT470264, WTCCCT449002, WTCCCT467316, WTC...</td>
      <td>23</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<hr>
<h3 id="The-problem-of-multiple-possible-input-types">The problem of multiple possible input types<a class="anchor-link" href="#The-problem-of-multiple-possible-input-types"> </a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The genetic file can be specified in multiple ways:</p>
<ul>
<li>Different formats (.gen, .bgen)</li>
<li>Split into multiple files for different phenotypes (case/control etc)</li>
<li>Split into one file per chromosome</li>
</ul>
<p>Also, the input files can be:</p>
<ul>
<li>Stored in local disk</li>
<li>In a compute cluster</li>
<li>In the cloud</li>
</ul>
<p>The library allows any combinations of these options. Let's look at the data catalog to see examples of these:</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#printing out the catalog</span>
<span class="o">!</span>cat conf/base/catalog_input/genetic_file.yaml
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>_MS_gen_file: &amp;MS_gen_file
    type: corradin_ovp_utils.datasets.OVPDataset.OVPDataset
    file_format: genetic_file.GenFileFormat
    load_args:
        prob_n_cols: 3
        initial_cols:
            - &#34;dashes&#34;
            - &#34;rsid&#34;
            - &#34;position&#34;
            - &#34;ref&#34;
            - &#34;alt&#34;
        rsid_col: &#34;rsid&#34;
        ref_col: &#34;ref&#34;
        alt_col: &#34;alt&#34;
        pandas_args:
            sep: &#34; &#34;
            header: null
            
        

genetic_file:
    &lt;&lt;: *MS_gen_file
    file_type: OVPDataset.CaseControlFilePathSchema
    file_path:
        case:
            folder: &#34;data/test_data/gen_file&#34;
            full_file_name: &#34;test_CASE_MS_chr22.gen&#34;
        control:
            folder: &#34;data/test_data/gen_file&#34;
            full_file_name: &#34;test_CONTROL_MS_chr22.gen&#34;

            
genetic_file_common_folder:
    &lt;&lt;: *MS_gen_file
    file_type: OVPDataset.CaseControlFilePathSchema
    file_path:
        common_folder: &#34;data/test_data/gen_file&#34;
        case:
            full_file_name: &#34;test_CASE_MS_chr22.gen&#34;
        control:
            full_file_name: &#34;test_CONTROL_MS_chr22.gen&#34;


genetic_file_single:
    &lt;&lt;: *MS_gen_file
    file_type: OVPDataset.SingleFilePathSchema
    file_path:
        folder: &#34;data/test_data/gen_file&#34;
        full_file_name: &#34;test_CASE_MS_chr22.gen&#34;


genetic_file_split_by_chrom:
    &lt;&lt;: *MS_gen_file
    file_type: OVPDataset.CaseControlFilePathSchema
    file_path:
        common_folder: &#34;data/test_data/gen_file&#34;
        case:
            split_by_chromosome: True
            full_file_name: &#34;test_CASE_MS_chr{chrom_num}.gen&#34;
        control:
            split_by_chromosome: True
            full_file_name: &#34;test_CONTROL_MS_chr{chrom_num}.gen&#34;</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">!</span>cat conf/base/catalog_input/sample_file.yaml
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>sample_file:
    type: corradin_ovp_utils.datasets.OVPDataset.OVPDataset
    file_type: OVPDataset.CaseControlFilePathSchema
    file_format: sample_file.SampleFileFormat
    load_args:
        sample_id_col: &#34;ID_2&#34;
        cov_cols: [&#34;sex&#34;]
        missing_col: &#34;missing&#34;
        pandas_args:
            sep: &#34; &#34;
            skiprows: [1] #2nd line of file is extra and should be discarded
    file_path:
        common_folder: &#34;data/test_data/sample_file&#34;
        case:
            full_file_name: &#34;MS_impute2_ALL_sample_out.tsv&#34;
        control:
            full_file_name: &#34;ALL_controls_58C_NBS_WTC2_impute2_sample_out.tsv&#34;</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

